<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>BatchSVG_analysis - humanHippocampus2024 Dataset</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">BatchSVG_analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./featureSelection.html" rel="" target="">
 <span class="menu-text">Why</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./spatialLIBD_spe.html" rel="" target="">
 <span class="menu-text">spatialLIBD</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./humanHippocampus2024_sub4_spe.html" rel="" target="" aria-current="page">
 <span class="menu-text">humanHippocampus2024</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#select-spatially-variable-genes" id="toc-select-spatially-variable-genes" class="nav-link" data-scroll-target="#select-spatially-variable-genes">Select Spatially Variable Genes</a></li>
  <li><a href="#implement-batchsvg" id="toc-implement-batchsvg" class="nav-link" data-scroll-target="#implement-batchsvg">Implement <code>BatchSVG</code></a></li>
  <li><a href="#cluster-results" id="toc-cluster-results" class="nav-link" data-scroll-target="#cluster-results">Cluster results</a></li>
  <li><a href="#r-session-information" id="toc-r-session-information" class="nav-link" data-scroll-target="#r-session-information"><code>R</code> session information</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">humanHippocampus2024 Dataset</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="introduction" class="level3">
<h3 class="anchored" data-anchor-id="introduction">Introduction</h3>
<p>In this section, we will include biased feature identification analysis using our developed method to show how <code>BatchSVG</code> helps to detect and visualize the biased features in SVGs. We will use the spatially-resolved transcriptomics (SRT) dataset from the <a href="https://bioconductor.org/packages/humanHippocampus2024/">humanHippocampus</a> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># library(humanHippocampus2024)</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ExperimentHub)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SpatialExperiment)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(PRECAST)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BatchSVG)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggspavis)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>humanHippocampus2024</code> dataset was collected from the adjacent tissue sections of the anterior human hippocampus (HPC) across ten adult neurotypical donors. The dataset is the <code>spatialExperiment</code> object generated and processed from the <a href="https://github.com/LieberInstitute/spatial_hpc">spatial_HPC</a> project. This dataset has 9 slides and 4 samples from each slide.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>ehub <span class="ot">&lt;-</span> <span class="fu">ExperimentHub</span>()</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> ehub[[<span class="st">"EH9605"</span>]]</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  31483 150917</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">colData</span>(spe)[,<span class="fu">c</span>(<span class="st">"sample_id"</span>,<span class="st">"slide"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               slide
sample_id       V10B01-085 V10B01-086 V11A20-297 V11L05-333 V11L05-335
  V10B01-085_A1       3654          0          0          0          0
  V10B01-085_B1       3263          0          0          0          0
  V10B01-085_C1       3738          0          0          0          0
  V10B01-085_D1       3348          0          0          0          0
  V10B01-086_A1          0       4655          0          0          0
  V10B01-086_B1          0       3964          0          0          0
  V10B01-086_C1          0       2606          0          0          0
  V10B01-086_D1          0       3389          0          0          0
  V11A20-297_A1          0          0       4432          0          0
  V11A20-297_B1          0          0       4659          0          0
  V11A20-297_C1          0          0       3429          0          0
  V11A20-297_D1          0          0       3412          0          0
  V11L05-333_A1          0          0          0       4969          0
  V11L05-333_B1          0          0          0       4985          0
  V11L05-333_C1          0          0          0       4548          0
  V11L05-333_D1          0          0          0       4938          0
  V11L05-335_A1          0          0          0          0       4607
  V11L05-335_B1          0          0          0          0       4737
  V11L05-335_C1          0          0          0          0       4887
  V11L05-335_D1          0          0          0          0       4483
  V11L05-336_A1          0          0          0          0          0
  V11L05-336_B1          0          0          0          0          0
  V11L05-336_C1          0          0          0          0          0
  V11L05-336_D1          0          0          0          0          0
  V11U08-081_A1          0          0          0          0          0
  V11U08-081_B1          0          0          0          0          0
  V11U08-081_C1          0          0          0          0          0
  V11U08-081_D1          0          0          0          0          0
  V11U08-084_A1          0          0          0          0          0
  V11U08-084_B1          0          0          0          0          0
  V11U08-084_C1          0          0          0          0          0
  V11U08-084_D1          0          0          0          0          0
  V12F14-051_A1          0          0          0          0          0
  V12F14-051_B1          0          0          0          0          0
  V12F14-051_C1          0          0          0          0          0
  V12F14-051_D1          0          0          0          0          0
               slide
sample_id       V11L05-336 V11U08-081 V11U08-084 V12F14-051
  V10B01-085_A1          0          0          0          0
  V10B01-085_B1          0          0          0          0
  V10B01-085_C1          0          0          0          0
  V10B01-085_D1          0          0          0          0
  V10B01-086_A1          0          0          0          0
  V10B01-086_B1          0          0          0          0
  V10B01-086_C1          0          0          0          0
  V10B01-086_D1          0          0          0          0
  V11A20-297_A1          0          0          0          0
  V11A20-297_B1          0          0          0          0
  V11A20-297_C1          0          0          0          0
  V11A20-297_D1          0          0          0          0
  V11L05-333_A1          0          0          0          0
  V11L05-333_B1          0          0          0          0
  V11L05-333_C1          0          0          0          0
  V11L05-333_D1          0          0          0          0
  V11L05-335_A1          0          0          0          0
  V11L05-335_B1          0          0          0          0
  V11L05-335_C1          0          0          0          0
  V11L05-335_D1          0          0          0          0
  V11L05-336_A1       4539          0          0          0
  V11L05-336_B1       4193          0          0          0
  V11L05-336_C1       4443          0          0          0
  V11L05-336_D1       4658          0          0          0
  V11U08-081_A1          0       4609          0          0
  V11U08-081_B1          0       3763          0          0
  V11U08-081_C1          0       3649          0          0
  V11U08-081_D1          0       4443          0          0
  V11U08-084_A1          0          0       4822          0
  V11U08-084_B1          0          0       4541          0
  V11U08-084_C1          0          0       4986          0
  V11U08-084_D1          0          0       4291          0
  V12F14-051_A1          0          0          0       3728
  V12F14-051_B1          0          0          0       3074
  V12F14-051_C1          0          0          0       4037
  V12F14-051_D1          0          0          0       4438</code></pre>
</div>
</div>
<p>We conducted data analysis using four samples from the raw data:</p>
<ul>
<li><p>V11L05-333_B1</p></li>
<li><p>V11L05-333_D1</p></li>
<li><p>V11L05-335_D1</p></li>
<li><p>V11L05-336_A1.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>fix_order <span class="ot">&lt;-</span> <span class="fu">distinct</span>(</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(spe)), slide, array, brnum, sample_id, </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    position, sex) <span class="sc">%&gt;%</span> </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(slide, array)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>sub4 <span class="ot">&lt;-</span> fix_order<span class="sc">$</span>sample_id[<span class="fu">c</span>(<span class="dv">14</span>,<span class="dv">16</span>, <span class="dv">20</span>,<span class="dv">21</span>)]</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>spe_sub4 <span class="ot">&lt;-</span> spe[,spe<span class="sc">$</span>sample_id <span class="sc">%in%</span> sub4]</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spe_sub4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 31483 18945</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sample_id vs. slide</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>table_data_slide <span class="ot">&lt;-</span> <span class="fu">table</span>(<span class="fu">colData</span>(spe_sub4)[, <span class="fu">c</span>(<span class="st">"sample_id"</span>, <span class="st">"slide"</span>)])</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>subset_table_slide <span class="ot">&lt;-</span> table_data_slide[<span class="fu">rownames</span>(table_data_slide) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"V11L05-333_B1"</span>, <span class="st">"V11L05-333_D1"</span>, <span class="st">"V11L05-335_D1"</span>, <span class="st">"V11L05-336_A1"</span>), ]</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>subset_table_slide_2 <span class="ot">&lt;-</span> subset_table_slide[,<span class="fu">colnames</span>(subset_table_slide) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"V11L05-333"</span>, <span class="st">"V11L05-335"</span>, <span class="st">"V11L05-336"</span>) ]</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>subset_table_slide_2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               slide
sample_id       V11L05-333 V11L05-335 V11L05-336
  V11L05-333_B1       4985          0          0
  V11L05-333_D1       4938          0          0
  V11L05-335_D1          0       4483          0
  V11L05-336_A1          0          0       4539</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sample_id vs. sex</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>table_data_sex <span class="ot">&lt;-</span> <span class="fu">table</span>(<span class="fu">colData</span>(spe_sub4)[, <span class="fu">c</span>(<span class="st">"sample_id"</span>, <span class="st">"sex"</span>)])</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>subset_table_sex <span class="ot">&lt;-</span> table_data_sex[<span class="fu">rownames</span>(table_data_sex) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"V11L05-333_B1"</span>, <span class="st">"V11L05-333_D1"</span>, <span class="st">"V11L05-335_D1"</span>, <span class="st">"V11L05-336_A1"</span>), ]</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>subset_table_sex</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               sex
sample_id          F    M
  V11L05-333_B1    0 4985
  V11L05-333_D1    0 4938
  V11L05-335_D1 4483    0
  V11L05-336_A1 4539    0</code></pre>
</div>
</div>
</section>
<section id="select-spatially-variable-genes" class="level3">
<h3 class="anchored" data-anchor-id="select-spatially-variable-genes">Select Spatially Variable Genes</h3>
<p>We used the spatially variable genes set generated from <a href="https://github.com/LieberInstitute/spatial_hpc">spatial_HPC</a> project, and the collaborators used <a href="(https://www.nature.com/articles/s41467-023-39748-z)">nnSVG</a> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">here</span>(<span class="st">"data/nnSVG_outs_HE_only.rda"</span>))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(res_ranks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 7028   36</code></pre>
</div>
</div>
<p>We refined our selection to include only the top 2,000 ranked features (rank <span class="math inline">\(\leq\)</span> 2000) and only genes that appear in more than one sample (n &gt; 1).</p>
<p>After applying these criteria, we obtain 2082 spatially variable genes across the four samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>res_df_sub <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames_to_column</span>(<span class="fu">as.data.frame</span>(res_ranks), var<span class="ot">&lt;-</span><span class="st">"gene_id"</span>), </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">colnames</span>(res_ranks), </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to=</span><span class="st">"sample_id"</span>, </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to=</span><span class="st">"rank"</span>, </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_drop_na=</span><span class="cn">TRUE</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>res_df_sub <span class="ot">&lt;-</span> <span class="fu">filter</span>(res_df_sub,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    sample_id <span class="sc">%in%</span> </span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">c</span>(<span class="st">"V11L05-333_B1"</span>, <span class="st">"V11L05-333_D1"</span>, <span class="st">"V11L05-335_D1"</span>, <span class="st">"V11L05-336_A1"</span>), </span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>    rank <span class="sc">&lt;=</span> <span class="dv">2000</span>) <span class="co"># top 2k sig features</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>svgs_sub4 <span class="ot">&lt;-</span> <span class="fu">group_by</span>(res_df_sub, gene_id) <span class="sc">|&gt;</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tally</span>() <span class="sc">|&gt;</span> </span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(n<span class="sc">&gt;</span><span class="dv">1</span>)</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(svgs_sub4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2082</code></pre>
</div>
</div>
<p>Finally, we obtained the subset spe object for our analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>spe_sub4 <span class="ot">&lt;-</span> spe_sub4[<span class="fu">rowData</span>(spe_sub4)<span class="sc">$</span>gene_id <span class="sc">%in%</span> svgs_sub4<span class="sc">$</span>gene_id,]</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(spe_sub4) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe_sub4)<span class="sc">$</span>gene_id</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>spe_sub4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SpatialExperiment 
dim: 2082 18945 
metadata(1): Obtained_from
assays(2): counts logcounts
rownames(2082): ENSG00000131584 ENSG00000175756 ... ENSG00000198695
  ENSG00000198727
rowData names(7): source type ... gene_type gene_search
colnames(18945): AAACAACGAATAGTTC-1_V11L05-333_B1
  AAACAAGTATCTCCCA-1_V11L05-333_B1 ... TTGTTTGTATTACACG-1_V11L05-336_A1
  TTGTTTGTGTAAATTC-1_V11L05-336_A1
colData names(150): sample_id in_tissue ... nmf99 nmf100
reducedDimNames(3): 10x_pca 10x_tsne 10x_umap
mainExpName: NULL
altExpNames(0):
spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres
imgData names(4): sample_id image_id data scaleFactor</code></pre>
</div>
</div>
</section>
<section id="implement-batchsvg" class="level3">
<h3 class="anchored" data-anchor-id="implement-batchsvg">Implement <code>BatchSVG</code></h3>
<section id="installation" class="level4">
<h4 class="anchored" data-anchor-id="installation">Installation</h4>
<p>(After accepted in <a href="https://bioconductor.org/">Bioconductor</a>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>)) {</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"BatchSVG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Install the development version from <a href="https://christinehou11.github.io/BatchSVG">GitHub</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install</span>(<span class="st">"christinehou11/BatchSVG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="batchsvgfeatureselect-perform-feature-selection" class="level4">
<h4 class="anchored" data-anchor-id="batchsvgfeatureselect-perform-feature-selection"><code>BatchSVG::featureSelect()</code>: Perform Feature Selection</h4>
<p>We applied <code>featureSelect()</code> to <code>spe_sub4</code> dataset while adjusting for the batch effect <em>sample_id</em> and <em>sex</em>. Different from the <code>spatialLIBD</code> dataset analysis, we showed how <code>BatchSVG</code> package can conduct the biased feature identification for multiple batch effects at the same time with specific thresholds for relative change in deviance and rank respectively.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>list_batch_df <span class="ot">&lt;-</span> <span class="fu">featureSelect</span>(<span class="at">input =</span> spe_sub4, </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">batch_effect =</span> <span class="fu">c</span>(<span class="st">"sample_id"</span>, <span class="st">"sex"</span>), <span class="at">VGs =</span> svgs_sub4<span class="sc">$</span>gene_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Running feature selection without batch...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Batch Effect: sample_id</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Running feature selection without batch...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating deviance and rank difference...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Batch Effect: sex</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Running feature selection without batch...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating deviance and rank difference...</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(list_batch_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "list"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(list_batch_df<span class="sc">$</span>sample_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default dev_sample_id
1 ENSG00000131584     ACAP3    16125.31         1262      15900.14
2 ENSG00000175756  AURKAIP1    17344.09         1060      17167.86
3 ENSG00000242485    MRPL20    17629.33         1023      17517.05
4 ENSG00000179403      VWA1    12860.93         1726      12825.66
5 ENSG00000160075     SSU72    16145.20         1255      16136.31
6 ENSG00000078369      GNB1    22402.83          516      22271.32
  rank_sample_id       d_diff nSD_dev_sample_id r_diff nSD_rank_sample_id
1           1269 0.0141612453       -0.09513109      7         0.16380252
2           1058 0.0102651525       -0.16945410     -2        -0.04680072
3           1004 0.0064098269       -0.24299943    -19        -0.44460683
4           1702 0.0027493688       -0.31282741    -24        -0.56160863
5           1220 0.0005506572       -0.35477067    -35        -0.81901258
6            497 0.0059049925       -0.25262980    -19        -0.44460683</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(list_batch_df<span class="sc">$</span>sex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default  dev_sex rank_sex
1 ENSG00000131584     ACAP3    16125.31         1262 16118.48     1250
2 ENSG00000175756  AURKAIP1    17344.09         1060 17247.44     1064
3 ENSG00000242485    MRPL20    17629.33         1023 17585.70     1013
4 ENSG00000179403      VWA1    12860.93         1726 12860.90     1709
5 ENSG00000160075     SSU72    16145.20         1255 16141.12     1243
6 ENSG00000078369      GNB1    22402.83          516 22314.17      509
        d_diff nSD_dev_sex r_diff nSD_rank_sex
1 4.234208e-04  -0.2615600    -12   -0.3080188
2 5.603690e-03  -0.1013769      4    0.1026729
3 2.480783e-03  -0.1979427    -10   -0.2566824
4 1.811106e-06  -0.2745969    -17   -0.4363600
5 2.527558e-04  -0.2668373    -12   -0.3080188
6 3.973515e-03  -0.1517848     -7   -0.1796776</code></pre>
</div>
</div>
</section>
<section id="batchsvgsvg_nsd-visualize-svg-selection-for-batch-effects" class="level4">
<h4 class="anchored" data-anchor-id="batchsvgsvg_nsd-visualize-svg-selection-for-batch-effects"><code>BatchSVG::svg_nSD()</code>: Visualize SVG Selection for Batch Effect(s)</h4>
<p>We utilized <code>svg_nSD()</code> function to generate visualizations for batch effects assessments in spatially variable genes (SVGs). The <code>svg_nSD()</code> function allows for multiple batch effects with corresponding <code>sd_interval_dev</code> and <code>sd_interval_rank</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">svg_nSD</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">sd_interval_dev =</span> <span class="fu">c</span>(<span class="dv">10</span>,<span class="dv">10</span>), <span class="at">sd_interval_rank =</span> <span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">6</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Figure 1a. Visualizations of nSD_dev and nSD_rank threshold selection for Sample ID</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>plots<span class="sc">$</span>sample_id</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/svg sample-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p><em>Figure 1b. Visualizations of nSD_dev and nSD_rank threshold selection for Sex</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>plots<span class="sc">$</span>sex</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/svg sex-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section>
<section id="batchsvgbiasdetect-identify-biased-genes" class="level4">
<h4 class="anchored" data-anchor-id="batchsvgbiasdetect-identify-biased-genes"><code>BatchSVG::biasDetect()</code>: Identify Biased Genes</h4>
<p>Using <code>nSD_svg()</code> function, we determined the nSD thresholds of relative change in deviance and rank for <code>sample_id</code> and <code>sex</code> respectively. Then, we applied <code>biasDetect()</code> function on the list of data frames obtained from <code>featureSelect()</code> function to identify and filter out biased genes. We used both thresholds (<code>threshold = "both"</code>) with the color customization.</p>
<ul>
<li><p>Sample ID</p>
<ul>
<li><code>nSD_dev = 10</code></li>
<li><code>nSD_rank = 5</code></li>
<li><code>plot_palette = "Blues"</code></li>
</ul></li>
<li><p>Sex</p>
<ul>
<li><code>nSD_dev = 10</code></li>
<li><code>nSD_rank = 6</code></li>
<li><code>plot_palette = "Purples"</code>.</li>
</ul></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>bias_both <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"both"</span>, <span class="at">nSD_dev =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">10</span>), <span class="at">nSD_rank =</span> <span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">6</span>), <span class="at">plot_palette =</span> <span class="fu">c</span>(<span class="st">"YlOrRd"</span>,<span class="st">"Purples"</span>), <span class="at">plot_point_size =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>), <span class="at">plot_point_shape =</span> <span class="fu">c</span>(<span class="dv">16</span>, <span class="dv">9</span>), <span class="at">plot_text_size =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="sample-id" class="level5">
<h5 class="anchored" data-anchor-id="sample-id">Sample ID</h5>
<p><em>Table 1. Outlier Genes defined by nSD_dev and nSD_rank for Sample ID</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>bias_both<span class="sc">$</span>sample_id<span class="sc">$</span>Table[,<span class="fu">c</span>(<span class="st">"gene_id"</span>,<span class="st">"gene_name"</span>,<span class="st">"nSD_bin_dev"</span>, <span class="st">"dev_outlier"</span>, <span class="st">"nSD_bin_rank"</span>, <span class="st">"rank_outlier"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           gene_id gene_name nSD_bin_dev dev_outlier nSD_bin_rank rank_outlier
1  ENSG00000177606       JUN      [0,10)       FALSE       [5,10)         TRUE
2  ENSG00000159388      BTG2      [0,10)       FALSE      [10,15)         TRUE
3  ENSG00000135625      EGR4      [0,10)       FALSE       [5,10)         TRUE
4  ENSG00000120738      EGR1      [0,10)       FALSE       [5,10)         TRUE
5  ENSG00000120129     DUSP1      [0,10)       FALSE      [10,15)         TRUE
6  ENSG00000204388    HSPA1B      [0,10)       FALSE      [10,15)         TRUE
7  ENSG00000130222   GADD45G      [0,10)       FALSE       [5,10)         TRUE
8  ENSG00000174576     NPAS4      [0,10)       FALSE       [5,10)         TRUE
9  ENSG00000123358     NR4A1      [0,10)       FALSE      [15,20]         TRUE
10 ENSG00000170345       FOS     [10,20)        TRUE        [0,5)        FALSE
11 ENSG00000256618  MTRNR2L1     [30,40]        TRUE       [5,10)         TRUE
12 ENSG00000099860   GADD45B      [0,10)       FALSE       [5,10)         TRUE
13 ENSG00000171223      JUNB      [0,10)       FALSE       [5,10)         TRUE
14 ENSG00000125740      FOSB      [0,10)       FALSE       [5,10)         TRUE
15 ENSG00000229807      XIST     [10,20)        TRUE      [15,20]         TRUE</code></pre>
</div>
</div>
<p><em>Figure 2a. Outlier Genes defined by nSD_dev and nSD_rank for Sample ID</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>bias_both<span class="sc">$</span>sample_id<span class="sc">$</span>Plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/sample id plot-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>Since we defined <code>threshold = "both"</code>, the <code>BatchSVG</code> package identified the biased features based on both <strong>dev_outlier</strong> and <strong>rank_outlier</strong>. The identified biased features may have</p>
<ul>
<li><p>dev_outlier = TRUE, rank_outlier = TRUE or</p></li>
<li><p>dev_outlier = FALSE, rank_outlier = TRUE or</p></li>
<li><p>dev_outlier = TRUE, rank_outlier = FALSE.</p></li>
</ul>
<p>For the case of dev_outlier = TRUE AND rank_outlier = TRUE, the <code>BatchSVG</code> package identified <strong>XIST</strong> and <strong>MTRNR2L1</strong> as bised features. For the case of dev_outlier = TRUE OR rank_outlier = TRUE, the <code>BatchSVG</code> package identified <strong>JUN</strong>, <strong>BTG2</strong>, <strong>EGR4</strong>, <strong>EGR1</strong>, <strong>DUSP1</strong>, <strong>HSPA1B</strong>, <strong>GADD45G</strong>, <strong>NPAS4</strong>, <strong>NR4A1</strong>, <strong>FOS</strong>, <strong>GADD45B</strong>, <strong>JUNB</strong>, and <strong>FOSB</strong>. We generated spatial expression plots for each identified genes to confirm that they are sample-biased features.</p>
<p><em>Figure 2b. Spatial expression plots for dev_outlier = TRUE AND rank_outlier = TRUE</em></p>
<div class="cell">
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/sample id both expression plots-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p><em>Figure 2c. Spatial expression plots for dev_outlier = TRUE OR rank_outlier = TRUE</em></p>
<div class="cell">
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/sample id either expression plots-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="sex" class="level5">
<h5 class="anchored" data-anchor-id="sex">Sex</h5>
<p><em>Table 2. Outlier Genes defined by nSD_dev and nSD_rank for Sex</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>bias_both<span class="sc">$</span>sex<span class="sc">$</span>Table[,<span class="fu">c</span>(<span class="st">"gene_id"</span>,<span class="st">"gene_name"</span>,<span class="st">"nSD_bin_dev"</span>, <span class="st">"dev_outlier"</span>, <span class="st">"nSD_bin_rank"</span>, <span class="st">"rank_outlier"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           gene_id gene_name nSD_bin_dev dev_outlier nSD_bin_rank rank_outlier
1  ENSG00000159388      BTG2      [0,10)       FALSE      [12,18)         TRUE
2  ENSG00000135625      EGR4      [0,10)       FALSE       [6,12)         TRUE
3  ENSG00000120738      EGR1      [0,10)       FALSE       [6,12)         TRUE
4  ENSG00000120129     DUSP1      [0,10)       FALSE      [12,18)         TRUE
5  ENSG00000204388    HSPA1B      [0,10)       FALSE       [6,12)         TRUE
6  ENSG00000130222   GADD45G      [0,10)       FALSE       [6,12)         TRUE
7  ENSG00000174576     NPAS4     [10,20)        TRUE        [0,6)        FALSE
8  ENSG00000123358     NR4A1     [10,20)        TRUE      [18,24]         TRUE
9  ENSG00000170345       FOS     [10,20)        TRUE        [0,6)        FALSE
10 ENSG00000256618  MTRNR2L1     [10,20)        TRUE        [0,6)        FALSE
11 ENSG00000099860   GADD45B      [0,10)       FALSE       [6,12)         TRUE
12 ENSG00000171223      JUNB      [0,10)       FALSE       [6,12)         TRUE
13 ENSG00000125740      FOSB      [0,10)       FALSE       [6,12)         TRUE
14 ENSG00000229807      XIST     [20,30]        TRUE      [12,18)         TRUE</code></pre>
</div>
</div>
<p><em>Figure 3a. Outlier Genes defined by nSD_dev and nSD_rank for Sex</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>bias_both<span class="sc">$</span>sex<span class="sc">$</span>Plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/sex plot-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>For the case of dev_outlier = TRUE AND rank_outlier = TRUE, the <code>BatchSVG</code> package identified <strong>NR4A1</strong> and <strong>XIST</strong> as bised features. For the case of dev_outlier = TRUE OR rank_outlier = TRUE, the <code>BatchSVG</code> package identified <strong>BTG2</strong>, <strong>EGR4</strong>, <strong>EGR1</strong>, <strong>DUSP1</strong>, <strong>HSPA1B</strong>, <strong>GADD45G</strong>, <strong>NPAS4</strong>, <strong>FOS</strong>, <strong>MTRNR2L1</strong>, <strong>GADD45B</strong>, <strong>JUNB</strong>, and <strong>FOSB</strong>. We generated spatial expression plots for each identified genes to confirm that they are sex-biased features.</p>
<p><em>Figure 3b. Spatial expression plots for dev_outlier = TRUE AND rank_outlier = TRUE</em></p>
<div class="cell">
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/sex both expression plots-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p><em>Figure 3c. Spatial expression plots for dev_outlier = TRUE OR rank_outlier = TRUE</em></p>
<div class="cell">
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/sex either expression plots-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
</section>
<section id="refine-svgs-by-removing-batch-affected-outliers" class="level4">
<h4 class="anchored" data-anchor-id="refine-svgs-by-removing-batch-affected-outliers">Refine SVGs by Removing Batch-Affected Outliers</h4>
<p>Finally, based on different batch effects, we were able to obtain a refined set of spatially variable genes (SVGs) by removing the identified outliers based on user-defined thresholds for <code>nSD_dev</code> and <code>nSD_rank</code>.</p>
<section id="sample-id-1" class="level5">
<h5 class="anchored" data-anchor-id="sample-id-1">Sample ID</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>bias_sample_id <span class="ot">&lt;-</span> bias_both<span class="sc">$</span>sample_id<span class="sc">$</span>Table</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>svgs_filt_sample_id <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(svgs_sub4<span class="sc">$</span>gene_id, bias_sample_id<span class="sc">$</span>gene_id)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>svgs_sub4_filt_sample_id <span class="ot">&lt;-</span> svgs_sub4[svgs_sub4<span class="sc">$</span>gene_id <span class="sc">%in%</span> svgs_filt_sample_id, ]</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(svgs_sub4_filt_sample_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2067</code></pre>
</div>
</div>
</section>
<section id="sex-1" class="level5">
<h5 class="anchored" data-anchor-id="sex-1">Sex</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>bias_sex <span class="ot">&lt;-</span> bias_both<span class="sc">$</span>sex<span class="sc">$</span>Table</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>svgs_filt_sex <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(svgs_sub4<span class="sc">$</span>gene_id, bias_sex<span class="sc">$</span>gene_id)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>svgs_sub4_filt_sex <span class="ot">&lt;-</span> svgs_sub4[svgs_sub4<span class="sc">$</span>gene_id <span class="sc">%in%</span> svgs_filt_sex, ]</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(svgs_sub4_filt_sex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2068</code></pre>
</div>
</div>
<p>After obtaining the refined set of SVGs for sample ID and sex respectively, these genes can be further analyzed using established spatial transcriptomics clustering algorithms to explore tissue layers and spatial organization.</p>
</section>
</section>
</section>
<section id="cluster-results" class="level3">
<h3 class="anchored" data-anchor-id="cluster-results">Cluster results</h3>
<p>We used the set of SVGs for sample ID to compare the cluster results before and after we remove the identified batch-related biased features using <code>BatchSVG</code> package. We performed <a href="https://www.nature.com/articles/s41467-023-35947-w">PRECAST</a> clustering to identify spatial domains after removing all sample-biased SVGs (n = 15) from the feature list.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat list</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>l2 <span class="ot">=</span> <span class="fu">unique</span>(spe_sub4<span class="sc">$</span>sample_id)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(l2) <span class="ot">=</span> l2</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>l2 <span class="ot">=</span> <span class="fu">lapply</span>(l2, <span class="cf">function</span>(x) spe_sub4[,<span class="fu">colData</span>(spe_sub4)<span class="sc">$</span>sample_id<span class="sc">==</span>x])</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>srt.sets <span class="ot">=</span> <span class="fu">lapply</span>(l2, <span class="cf">function</span>(x) {</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(<span class="fu">counts</span>(x)) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">colData</span>(x))</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colData</span>(x)<span class="sc">$</span>col <span class="ot">&lt;-</span> x<span class="sc">$</span>array_col</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colData</span>(x)<span class="sc">$</span>row <span class="ot">&lt;-</span> x<span class="sc">$</span>array_row</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>  count <span class="ot">&lt;-</span> <span class="fu">counts</span>(x)</span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>  a1 <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(count, <span class="at">assay =</span> <span class="st">"RNA"</span>, <span class="at">min.features =</span> <span class="dv">0</span>, <span class="at">min.cells =</span> <span class="dv">0</span>)</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">CreateSeuratObject</span>(a1, <span class="at">meta.data =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(x)))</span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a><span class="co">#run precast</span></span>
<span id="cb48-16"><a href="#cb48-16" aria-hidden="true" tabindex="-1"></a>preobj_old <span class="ot">&lt;-</span> <span class="fu">CreatePRECASTObject</span>(<span class="at">seuList =</span> srt.sets,</span>
<span id="cb48-17"><a href="#cb48-17" aria-hidden="true" tabindex="-1"></a>                              <span class="at">customGenelist=</span><span class="fu">rownames</span>(spe_sub4),</span>
<span id="cb48-18"><a href="#cb48-18" aria-hidden="true" tabindex="-1"></a>                              <span class="at">premin.spots=</span><span class="dv">0</span>, <span class="at">premin.features=</span><span class="dv">0</span>, <span class="at">postmin.spots=</span><span class="dv">0</span>, <span class="at">postmin.features=</span><span class="dv">0</span>)</span>
<span id="cb48-19"><a href="#cb48-19" aria-hidden="true" tabindex="-1"></a>PRECASTObj_old <span class="ot">&lt;-</span> <span class="fu">AddAdjList</span>(preobj_old, <span class="at">platform =</span> <span class="st">"Visium"</span>)</span>
<span id="cb48-20"><a href="#cb48-20" aria-hidden="true" tabindex="-1"></a>PRECASTObj_old <span class="ot">&lt;-</span> <span class="fu">AddParSetting</span>(PRECASTObj_old, <span class="at">maxIter =</span> <span class="dv">20</span>, <span class="at">verbose =</span> <span class="cn">TRUE</span>, <span class="at">Sigma_equal=</span><span class="cn">FALSE</span>, <span class="at">coreNum=</span><span class="dv">12</span>)</span>
<span id="cb48-21"><a href="#cb48-21" aria-hidden="true" tabindex="-1"></a>PRECASTObj_old <span class="ot">&lt;-</span> <span class="fu">PRECAST</span>(PRECASTObj, <span class="at">K=</span><span class="dv">7</span>)</span>
<span id="cb48-22"><a href="#cb48-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-23"><a href="#cb48-23" aria-hidden="true" tabindex="-1"></a><span class="co">#consolidate/ reformat results</span></span>
<span id="cb48-24"><a href="#cb48-24" aria-hidden="true" tabindex="-1"></a>PRECASTObj_old <span class="ot">&lt;-</span> <span class="fu">SelectModel</span>(PRECASTObj_old, <span class="at">criteria=</span><span class="st">"MBIC"</span>)</span>
<span id="cb48-25"><a href="#cb48-25" aria-hidden="true" tabindex="-1"></a>seuInt_old <span class="ot">&lt;-</span> <span class="fu">IntegrateSpaData</span>(PRECASTObj_old, <span class="at">species =</span> <span class="st">"Human"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="refined-svgs" class="level4">
<h4 class="anchored" data-anchor-id="refined-svgs">Refined SVGs</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co">#remove biased genes</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>svgs_filt <span class="ot">=</span> <span class="fu">setdiff</span>(<span class="fu">rownames</span>(spe_sub4), bias_sample_id<span class="sc">$</span>gene_id) <span class="co"># 2067</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="co">#run precast</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>preobj_new <span class="ot">&lt;-</span> <span class="fu">CreatePRECASTObject</span>(<span class="at">seuList =</span> srt.sets,</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">customGenelist=</span>svgs_filt,</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">premin.spots=</span><span class="dv">0</span>, <span class="at">premin.features=</span><span class="dv">0</span>, <span class="at">postmin.spots=</span><span class="dv">0</span>, <span class="at">postmin.features=</span><span class="dv">0</span>)</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>PRECASTObj_new <span class="ot">&lt;-</span> <span class="fu">AddAdjList</span>(preobj_new, <span class="at">platform =</span> <span class="st">"Visium"</span>)</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>PRECASTObj_new <span class="ot">&lt;-</span> <span class="fu">AddParSetting</span>(PRECASTObj_new, <span class="at">maxIter =</span> <span class="dv">20</span>, <span class="at">verbose =</span> <span class="cn">TRUE</span>, <span class="at">Sigma_equal=</span><span class="cn">FALSE</span>, <span class="at">coreNum=</span><span class="dv">12</span>)</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>PRECASTObj_new <span class="ot">&lt;-</span> <span class="fu">PRECAST</span>(PRECASTObj_new, <span class="at">K=</span><span class="dv">7</span>)</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a><span class="co">#consolidate/ reformat results</span></span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>PRECASTObj_new <span class="ot">&lt;-</span> <span class="fu">SelectModel</span>(PRECASTObj_new, <span class="at">criteria=</span><span class="st">"MBIC"</span>)</span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>seuInt_new <span class="ot">&lt;-</span> <span class="fu">IntegrateSpaData</span>(PRECASTObj_new, <span class="at">species =</span> <span class="st">"Human"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plots-comparison" class="level4">
<h4 class="anchored" data-anchor-id="plots-comparison">Plots comparison</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(spe_sub4)[,<span class="fu">c</span>(<span class="st">"sample_id"</span>,<span class="st">"slide"</span>,<span class="st">"cluster"</span>)])</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>spe_sub4<span class="sc">$</span>precast_k7 <span class="ot">=</span> clusters1<span class="sc">$</span>cluster</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>spe_sub4<span class="sc">$</span>precast_k7_ordered <span class="ot">=</span> <span class="fu">factor</span>(spe_sub4<span class="sc">$</span>precast_k7, <span class="at">levels=</span><span class="fu">c</span>(<span class="dv">7</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">1</span>,<span class="dv">6</span>,<span class="dv">5</span>,<span class="dv">4</span>), </span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>                                <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"WM"</span>,<span class="st">"WM (2)"</span>,<span class="st">"SR/SL"</span>,<span class="st">"CA1"</span>,<span class="st">"CA3"</span>,<span class="st">"DG GCL"</span>,<span class="st">"DG ML"</span>))</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>spe_sub4<span class="sc">$</span>precast_k7_nobias <span class="ot">=</span> clusters2<span class="sc">$</span>cluster</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>spe_sub4<span class="sc">$</span>precast_k7_nobias_ordered<span class="ot">=</span> <span class="fu">factor</span>(spe_sub4<span class="sc">$</span>precast_k7_nobias, <span class="at">levels=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">7</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">4</span>,<span class="dv">3</span>),</span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>                                <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"WM"</span>,<span class="st">"SR/SL"</span>,<span class="st">"CA1"</span>,<span class="st">"CA1 (2)"</span>,<span class="st">"CA3"</span>,<span class="st">"DG GCL"</span>,<span class="st">"DG ML"</span>))</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>l2 <span class="ot">=</span> <span class="fu">unique</span>(spe_sub4<span class="sc">$</span>sample_id)</span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(l2) <span class="ot">=</span> l2</span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>l2 <span class="ot">=</span> <span class="fu">lapply</span>(l2, <span class="cf">function</span>(x) spe_sub4[,<span class="fu">colData</span>(spe_sub4)<span class="sc">$</span>sample_id<span class="sc">==</span>x])</span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-15"><a href="#cb50-15" aria-hidden="true" tabindex="-1"></a>col.pal1 <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"#1f77b4FF"</span>,<span class="st">"#aec7e8FF"</span>,<span class="st">"#ffbb78FF"</span>,<span class="st">"#2ca02cFF"</span>,<span class="st">"#ff7f0eFF"</span>,<span class="st">"#d62728FF"</span>,<span class="st">"#ff9896FF"</span>)</span>
<span id="cb50-16"><a href="#cb50-16" aria-hidden="true" tabindex="-1"></a>col.pal2 <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"#1f77b4FF"</span>,<span class="st">"#ffbb78FF"</span>,<span class="st">"#2ca02cFF"</span>,<span class="st">"#98df8aFF"</span>,<span class="st">"#ff7f0eFF"</span>,<span class="st">"#d62728FF"</span>,<span class="st">"#ff9896FF"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Comparing the clustering results, the refined SVGs led to a more coherent and biologically meaningful spatial domain segmentation. Specifically, in the original clustering, an additional “WM (2)” cluster appeared, likely due to batch-associated biases. After applying <code>BatchSVG</code> , this extra cluster was removed, and the CA1 region became more distinct, resolving ambiguity in its boundaries. The refined clustering thus provides improved spatial structure alignment, demonstrating the effectiveness of BatchSVG in enhancing spatial domain identification.</p>
<p><em>Figure 4a. PRECAST clusters: 2082 input features, K=7</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>c1 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(l2), <span class="cf">function</span>(x) {</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plotSpots</span>(l2[[x]], <span class="at">annotate=</span><span class="st">"precast_k7_ordered"</span>, <span class="at">point_size=</span>.<span class="dv">3</span>)<span class="sc">+</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">color=</span><span class="st">"clus"</span>)<span class="sc">+</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values=</span>col.pal1)<span class="sc">+</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">8</span>))</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>PRECAST<span class="sc">::</span><span class="fu">drawFigs</span>(c1, <span class="at">layout.dim =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">4</span>), <span class="at">common.legend =</span> <span class="cn">TRUE</span>, <span class="at">legend.position =</span> <span class="st">"right"</span>, <span class="at">align =</span> <span class="st">"h"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/cluster result old-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><em>Figure 4b. PRECAST clusters: 2067 input features (SVGs minus 15 biased), K=7</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>c2 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(l2), <span class="cf">function</span>(x) {</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plotSpots</span>(l2[[x]], <span class="at">annotate=</span><span class="st">"precast_k7_nobias_ordered"</span>, <span class="at">point_size=</span>.<span class="dv">3</span>)<span class="sc">+</span></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">color=</span><span class="st">"clus"</span>)<span class="sc">+</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values=</span>col.pal2)<span class="sc">+</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">8</span>))</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>PRECAST<span class="sc">::</span><span class="fu">drawFigs</span>(c2, <span class="at">layout.dim =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">4</span>), <span class="at">common.legend =</span> <span class="cn">TRUE</span>, <span class="at">legend.position =</span> <span class="st">"right"</span>, <span class="at">align =</span> <span class="st">"h"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/cluster result new-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="r-session-information" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="r-session-information"><code>R</code> session information</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Session info</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] humanHippocampus2024_0.99.8 scater_1.34.1              
 [3] scuttle_1.16.0              gridExtra_2.3              
 [5] ggspavis_1.12.0             ggplot2_3.5.1              
 [7] BatchSVG_0.99.8             Seurat_5.2.1               
 [9] SeuratObject_5.0.2          sp_2.2-0                   
[11] PRECAST_1.6.5               gtools_3.9.5               
[13] here_1.0.1                  cowplot_1.1.3              
[15] tibble_3.2.1                dplyr_1.1.4                
[17] tidyr_1.3.1                 SpatialExperiment_1.16.0   
[19] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
[21] Biobase_2.66.0              GenomicRanges_1.58.0       
[23] GenomeInfoDb_1.42.3         IRanges_2.40.1             
[25] S4Vectors_0.44.0            MatrixGenerics_1.18.1      
[27] matrixStats_1.5.0           ExperimentHub_2.14.0       
[29] AnnotationHub_3.14.0        BiocFileCache_2.14.0       
[31] dbplyr_2.5.0                BiocGenerics_0.52.0        

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22        splines_4.4.2           later_1.4.1            
  [4] bitops_1.0-9            filelock_1.0.3          polyclip_1.10-7        
  [7] fastDummies_1.7.5       lifecycle_1.0.4         rstatix_0.7.2          
 [10] rprojroot_2.0.4         globals_0.16.3          lattice_0.22-6         
 [13] MASS_7.3-65             backports_1.5.0         magrittr_2.0.3         
 [16] plotly_4.10.4           rmarkdown_2.29          yaml_2.3.10            
 [19] httpuv_1.6.15           sctransform_0.4.1       ggside_0.3.1           
 [22] spam_2.11-1             spatstat.sparse_3.1-0   reticulate_1.41.0.1    
 [25] pbapply_1.7-2           DBI_1.2.3               RColorBrewer_1.1-3     
 [28] abind_1.4-8             zlibbioc_1.52.0         Rtsne_0.17             
 [31] purrr_1.0.4             RCurl_1.98-1.17         rappdirs_0.3.3         
 [34] GenomeInfoDbData_1.2.13 ggrepel_0.9.6           scry_1.18.0            
 [37] irlba_2.3.5.1           listenv_0.9.1           spatstat.utils_3.1-3   
 [40] goftest_1.2-3           RSpectra_0.16-2         spatstat.random_3.3-3  
 [43] fitdistrplus_1.2-2      parallelly_1.43.0       codetools_0.2-20       
 [46] DelayedArray_0.32.0     tidyselect_1.2.1        UCSC.utils_1.2.0       
 [49] farver_2.1.2            viridis_0.6.5           ScaledMatrix_1.14.0    
 [52] spatstat.explore_3.4-2  jsonlite_1.9.1          BiocNeighbors_2.0.1    
 [55] Formula_1.2-5           progressr_0.15.1        ggridges_0.5.6         
 [58] survival_3.8-3          tools_4.4.2             ica_1.0-3              
 [61] Rcpp_1.0.14             glue_1.8.0              SparseArray_1.6.2      
 [64] xfun_0.51               ggthemes_5.1.0          withr_3.0.2            
 [67] BiocManager_1.30.25     fastmap_1.2.0           rsvd_1.0.5             
 [70] digest_0.6.37           R6_2.6.1                mime_0.13              
 [73] colorspace_2.1-1        scattermore_1.2         tensor_1.5             
 [76] spatstat.data_3.1-6     RSQLite_2.3.9           generics_0.1.3         
 [79] data.table_1.17.0       httr_1.4.7              htmlwidgets_1.6.4      
 [82] S4Arrays_1.6.0          uwot_0.2.3              pkgconfig_2.0.3        
 [85] gtable_0.3.6            blob_1.2.4              lmtest_0.9-40          
 [88] XVector_0.46.0          htmltools_0.5.8.1       carData_3.0-5          
 [91] dotCall64_1.2           scales_1.3.0            png_0.1-8              
 [94] spatstat.univar_3.1-2   knitr_1.50              rstudioapi_0.17.1      
 [97] reshape2_1.4.4          rjson_0.2.23            nlme_3.1-167           
[100] curl_6.2.2              cachem_1.1.0            zoo_1.8-13             
[103] stringr_1.5.1           BiocVersion_3.20.0      KernSmooth_2.23-26     
[106] vipor_0.4.7             miniUI_0.1.1.1          GiRaF_1.0.1            
[109] AnnotationDbi_1.68.0    pillar_1.10.1           grid_4.4.2             
[112] vctrs_0.6.5             RANN_2.6.2              ggpubr_0.6.0           
[115] promises_1.3.2          car_3.1-3               BiocSingular_1.22.0    
[118] beachmat_2.22.0         DR.SC_3.4               xtable_1.8-4           
[121] cluster_2.1.8.1         beeswarm_0.4.0          evaluate_1.0.3         
[124] magick_2.8.6            cli_3.6.4               compiler_4.4.2         
[127] rlang_1.1.5             crayon_1.5.3            ggsignif_0.6.4         
[130] future.apply_1.11.3     labeling_0.4.3          mclust_6.1.1           
[133] ggbeeswarm_0.7.2        plyr_1.8.9              stringi_1.8.4          
[136] BiocParallel_1.40.0     viridisLite_0.4.2       deldir_2.0-4           
[139] munsell_0.5.1           Biostrings_2.74.1       lazyeval_0.2.2         
[142] spatstat.geom_3.3-6     CompQuadForm_1.4.3      Matrix_1.7-3           
[145] RcppHNSW_0.6.0          patchwork_1.3.0         bit64_4.6.0-1          
[148] future_1.34.0           KEGGREST_1.46.0         shiny_1.10.0           
[151] ROCR_1.0-11             broom_1.0.7             igraph_2.1.4           
[154] memoise_2.0.1           bit_4.6.0              </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>