<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>BatchSVG_analysis - humanHippocampus2024 Dataset</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">BatchSVG_analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./featureSelection.html" rel="" target="">
 <span class="menu-text">Why</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./dataDrivenThreshold.html" rel="" target="">
 <span class="menu-text">How</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./humanHippocampus2024_sub4_spe.html" rel="" target="" aria-current="page">
 <span class="menu-text">analysis_1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./spatialLIBD_spe.html" rel="" target="">
 <span class="menu-text">analysis_2</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#select-spatially-variable-genes" id="toc-select-spatially-variable-genes" class="nav-link" data-scroll-target="#select-spatially-variable-genes">Select Spatially Variable Genes</a></li>
  <li><a href="#implement-batchsvg" id="toc-implement-batchsvg" class="nav-link" data-scroll-target="#implement-batchsvg">Implement <code>BatchSVG</code></a></li>
  <li><a href="#compare-clustering-performances" id="toc-compare-clustering-performances" class="nav-link" data-scroll-target="#compare-clustering-performances">Compare clustering performances</a></li>
  <li><a href="#r-session-information" id="toc-r-session-information" class="nav-link" data-scroll-target="#r-session-information"><code>R</code> session information</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">humanHippocampus2024 Dataset</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="introduction" class="level3">
<h3 class="anchored" data-anchor-id="introduction">Introduction</h3>
<p>In this section, we will include biased feature identification analysis using our developed method to show how <code>BatchSVG</code> helps to detect and visualize the biased features in SVGs. We will use the spatially-resolved transcriptomics (SRT) dataset from the <a href="https://bioconductor.org/packages/humanHippocampus2024/">humanHippocampus</a> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># library(humanHippocampus2024)</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ExperimentHub)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SpatialExperiment)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(PRECAST)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BatchSVG)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggspavis)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>humanHippocampus2024</code> dataset was collected from the adjacent tissue sections of the anterior human hippocampus (HPC) across ten adult neurotypical donors. The dataset is the <code>spatialExperiment</code> object generated and processed from the <a href="https://github.com/LieberInstitute/spatial_hpc">spatial_HPC</a> project.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>ehub <span class="ot">&lt;-</span> <span class="fu">ExperimentHub</span>()</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>spe <span class="ot">&lt;-</span> ehub[[<span class="st">"EH9605"</span>]]</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  31483 150917</code></pre>
</div>
</div>
<p>We conducted data analysis using four samples from the raw data:</p>
<ul>
<li><p>V11L05-333_B1</p></li>
<li><p>V11L05-333_D1</p></li>
<li><p>V11L05-335_D1</p></li>
<li><p>V11L05-336_A1.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>fix_order <span class="ot">&lt;-</span> <span class="fu">distinct</span>(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(spe)), slide, array, brnum, sample_id, </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    position, sex) <span class="sc">%&gt;%</span> </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(slide, array)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>sub4 <span class="ot">&lt;-</span> fix_order<span class="sc">$</span>sample_id[<span class="fu">c</span>(<span class="dv">14</span>,<span class="dv">16</span>, <span class="dv">20</span>,<span class="dv">21</span>)]</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>spe_sub4 <span class="ot">&lt;-</span> spe[,spe<span class="sc">$</span>sample_id <span class="sc">%in%</span> sub4]</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spe_sub4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 31483 18945</code></pre>
</div>
</div>
</section>
<section id="select-spatially-variable-genes" class="level3">
<h3 class="anchored" data-anchor-id="select-spatially-variable-genes">Select Spatially Variable Genes</h3>
<p>We used the spatially variable genes set generated from <a href="https://github.com/LieberInstitute/spatial_hpc">spatial_HPC</a> project, and the collaborators used <a href="(https://www.nature.com/articles/s41467-023-39748-z)">nnSVG</a> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">here</span>(<span class="st">"data/nnSVG_outs_HE_only.rda"</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(res_ranks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 7028   36</code></pre>
</div>
</div>
<p>We refined our selection to include only the top 2,000 ranked features (rank <span class="math inline">\(\leq\)</span> 2000) and only genes that appear in more than one sample (n &gt; 1).</p>
<p>After applying these criteria, we obtain 2,082 spatially variable genes across the four samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>res_df_sub <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames_to_column</span>(<span class="fu">as.data.frame</span>(res_ranks), var<span class="ot">&lt;-</span><span class="st">"gene_id"</span>), </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">colnames</span>(res_ranks), </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to=</span><span class="st">"sample_id"</span>, </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to=</span><span class="st">"rank"</span>, </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_drop_na=</span><span class="cn">TRUE</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>res_df_sub <span class="ot">&lt;-</span> <span class="fu">filter</span>(res_df_sub,</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    sample_id <span class="sc">%in%</span> </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">c</span>(<span class="st">"V11L05-333_B1"</span>, <span class="st">"V11L05-333_D1"</span>, <span class="st">"V11L05-335_D1"</span>, <span class="st">"V11L05-336_A1"</span>), </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    rank <span class="sc">&lt;=</span> <span class="dv">2000</span>) <span class="co"># top 2k sig features</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>svgs_sub4 <span class="ot">&lt;-</span> <span class="fu">group_by</span>(res_df_sub, gene_id) <span class="sc">|&gt;</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tally</span>() <span class="sc">|&gt;</span> </span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(n<span class="sc">&gt;</span><span class="dv">1</span>)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(svgs_sub4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2082</code></pre>
</div>
</div>
<p>Finally, we obtained the subset spe object for our analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>spe_sub4 <span class="ot">&lt;-</span> spe_sub4[<span class="fu">rowData</span>(spe_sub4)<span class="sc">$</span>gene_id <span class="sc">%in%</span> svgs_sub4<span class="sc">$</span>gene_id,]</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(spe_sub4) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe_sub4)<span class="sc">$</span>gene_id</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>spe_sub4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SpatialExperiment 
dim: 2082 18945 
metadata(1): Obtained_from
assays(2): counts logcounts
rownames(2082): ENSG00000131584 ENSG00000175756 ... ENSG00000198695
  ENSG00000198727
rowData names(7): source type ... gene_type gene_search
colnames(18945): AAACAACGAATAGTTC-1_V11L05-333_B1
  AAACAAGTATCTCCCA-1_V11L05-333_B1 ... TTGTTTGTATTACACG-1_V11L05-336_A1
  TTGTTTGTGTAAATTC-1_V11L05-336_A1
colData names(150): sample_id in_tissue ... nmf99 nmf100
reducedDimNames(3): 10x_pca 10x_tsne 10x_umap
mainExpName: NULL
altExpNames(0):
spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres
imgData names(4): sample_id image_id data scaleFactor</code></pre>
</div>
</div>
</section>
<section id="implement-batchsvg" class="level3">
<h3 class="anchored" data-anchor-id="implement-batchsvg">Implement <code>BatchSVG</code></h3>
<section id="installation" class="level4">
<h4 class="anchored" data-anchor-id="installation">Installation</h4>
<p>(After accepted in <a href="https://bioconductor.org/">Bioconductor</a>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>)) {</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"BatchSVG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Install the development version from <a href="https://christinehou11.github.io/BatchSVG">GitHub</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install</span>(<span class="st">"christinehou11/BatchSVG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="batchsvgfeatureselect-perform-feature-selection" class="level4">
<h4 class="anchored" data-anchor-id="batchsvgfeatureselect-perform-feature-selection"><code>BatchSVG::featureSelect()</code>: Perform Feature Selection</h4>
<p>We applied <code>featureSelect()</code> to <code>spe_sub4</code> dataset while adjusting for the batch effect <em>sample_id</em> and <em>sex</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>list_batch_df <span class="ot">&lt;-</span> <span class="fu">featureSelect</span>(<span class="at">input =</span> spe_sub4, </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">batch_effect =</span> <span class="fu">c</span>(<span class="st">"sample_id"</span>, <span class="st">"sex"</span>), <span class="at">VGs =</span> svgs_sub4<span class="sc">$</span>gene_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Running feature selection without batch...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Batch Effect: sample_id</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Running feature selection without batch...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating deviance and rank difference...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Batch Effect: sex</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Running feature selection without batch...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating deviance and rank difference...</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(list_batch_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "list"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(list_batch_df<span class="sc">$</span>sample_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default dev_sample_id
1 ENSG00000131584     ACAP3    16125.31         1262      15900.14
2 ENSG00000175756  AURKAIP1    17344.09         1060      17167.86
3 ENSG00000242485    MRPL20    17629.33         1023      17517.05
4 ENSG00000179403      VWA1    12860.93         1726      12825.66
5 ENSG00000160075     SSU72    16145.20         1255      16136.31
6 ENSG00000078369      GNB1    22402.83          516      22271.32
  rank_sample_id       d_diff nSD_dev_sample_id r_diff nSD_rank_sample_id
1           1269 0.0141612453       -0.09513109      7         0.16380252
2           1058 0.0102651525       -0.16945410     -2        -0.04680072
3           1004 0.0064098269       -0.24299943    -19        -0.44460683
4           1702 0.0027493688       -0.31282741    -24        -0.56160863
5           1220 0.0005506572       -0.35477067    -35        -0.81901258
6            497 0.0059049925       -0.25262980    -19        -0.44460683</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(list_batch_df<span class="sc">$</span>sex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default  dev_sex rank_sex
1 ENSG00000131584     ACAP3    16125.31         1262 16118.48     1250
2 ENSG00000175756  AURKAIP1    17344.09         1060 17247.44     1064
3 ENSG00000242485    MRPL20    17629.33         1023 17585.70     1013
4 ENSG00000179403      VWA1    12860.93         1726 12860.90     1709
5 ENSG00000160075     SSU72    16145.20         1255 16141.12     1243
6 ENSG00000078369      GNB1    22402.83          516 22314.17      509
        d_diff nSD_dev_sex r_diff nSD_rank_sex
1 4.234208e-04  -0.2615600    -12   -0.3080188
2 5.603690e-03  -0.1013769      4    0.1026729
3 2.480783e-03  -0.1979427    -10   -0.2566824
4 1.811106e-06  -0.2745969    -17   -0.4363600
5 2.527558e-04  -0.2668373    -12   -0.3080188
6 3.973515e-03  -0.1517848     -7   -0.1796776</code></pre>
</div>
</div>
</section>
<section id="batchsvgsvg_nsd-visualize-svg-selection-for-batch-effects" class="level4">
<h4 class="anchored" data-anchor-id="batchsvgsvg_nsd-visualize-svg-selection-for-batch-effects"><code>BatchSVG::svg_nSD()</code>: Visualize SVG Selection for Batch Effect(s)</h4>
<p>We utilized <code>svg_nSD()</code> function to generate visualizations for batch effects assessments in spatially variable genes (SVGs). The <code>svg_nSD()</code> function allows for multiple batch effects with corresponding <code>sd_interval_dev</code> and <code>sd_interval_rank</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">svg_nSD</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">sd_interval_dev =</span> <span class="fu">c</span>(<span class="dv">10</span>,<span class="dv">10</span>), <span class="at">sd_interval_rank =</span> <span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">6</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>plots<span class="sc">$</span>sample_id</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/svg sample-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>plots<span class="sc">$</span>sex</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/svg sex-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section>
<section id="batchsvgbiasdetect-identify-biased-genes" class="level4">
<h4 class="anchored" data-anchor-id="batchsvgbiasdetect-identify-biased-genes"><code>BatchSVG::biasDetect()</code>: Identify Biased Genes</h4>
<p>The function <code>biasDetect()</code> is designed to identify and filter out biased genes across different batch effects. Using threshold values selected from the visualization results generated by <code>svg_nSD()</code>, this function systematically detects outliers that exceed a specified normalized standard deviation (nSD) threshold in either relative deviance change, rank difference, or both.</p>
<p>The function outputs visualizations comparing deviance and rank values with and without batch effects. Genes with high deviations, highlighted in color, are identified as potentially biased and can be excluded based on the selected nSD thresholds.</p>
<p>The function offers flexibility in customizing the plot aesthetics, allowing users to adjust the data point size (<strong>plot_point_size</strong>), shape (<strong>plot_point_shape</strong>), annotated text size (<strong>plot_text_size</strong>), and data point color pallete (<strong>plot_pallete</strong>). Default values are provided for these parameters if not specified. Users should refer to <a href="https://ggplot2.tidyverse.org/index.html">ggplot2</a> aesthetic guidelines to ensure appropriate values are assigned for each parameter.</p>
<p>We will use <code>nSD_dev = 7</code> and <code>nSD_rank = 6</code> as the example. The user should adjust the value based on their dataset features.</p>
<p><strong>Usage of Different Threshold Options</strong></p>
<ul>
<li><code>threshold = "dev"</code>: Filters biased genes based only on the relative change in deviance. Genes with deviance changes exceeding the specified <code>nSD_dev</code> threshold are identified as batch-affected and can be removed.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>bias_dev <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"dev"</span>, <span class="at">nSD_dev =</span> <span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bias_dev<span class="sc">$</span>sample_id<span class="sc">$</span>Table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default dev_sample_id
1 ENSG00000174576     NPAS4    35003.31          125     24629.414
2 ENSG00000123358     NR4A1    23299.81          457     16115.928
3 ENSG00000170345       FOS    42305.65           73     27146.089
4 ENSG00000256618  MTRNR2L1    69206.34           28     24876.086
5 ENSG00000118271       TTR  4719046.58            1   3292127.945
6 ENSG00000229807      XIST    15223.50         1408      8819.689
  rank_sample_id    d_diff nSD_dev_sample_id r_diff nSD_rank_sample_id
1            363 0.4211996          7.669653    238           5.569286
2           1226 0.4457631          8.138234    769          17.994876
3            263 0.5584435         10.287758    190           4.446068
4            351 1.7820430         33.629502    323           7.558316
5              1 0.4334335          7.903030      0           0.000000
6           2050 0.7260812         13.485664    642          15.023031
  nSD_bin_dev dev_outlier
1      [7,14)        TRUE
2      [7,14)        TRUE
3      [7,14)        TRUE
4     [28,35]        TRUE
5      [7,14)        TRUE
6      [7,14)        TRUE</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>bias_dev<span class="sc">$</span>sample_id<span class="sc">$</span>Plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/bias detect dev plot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can change the data point size using <strong>plot_point_size</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># size default = 3</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>bias_dev_size <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"dev"</span>, <span class="at">nSD_dev =</span> <span class="dv">7</span>, <span class="at">plot_point_size =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>))</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(bias_dev_size<span class="sc">$</span>sample_id<span class="sc">$</span>Plot,bias_dev_size<span class="sc">$</span>sex<span class="sc">$</span>Plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/size change-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<ul>
<li><code>threshold = "rank"</code>: Identifies biased genes based solely on rank difference. Genes with rank shifts exceeding <code>nSD_rank</code> are considered biased.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>bias_rank <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"rank"</span>, <span class="at">nSD_rank =</span> <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bias_rank<span class="sc">$</span>sex<span class="sc">$</span>Table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default  dev_sex rank_sex
1 ENSG00000159388      BTG2    18311.28          926 14257.70     1543
2 ENSG00000135625      EGR4    20336.84          705 17851.54      972
3 ENSG00000120738      EGR1    19882.54          752 17444.96     1037
4 ENSG00000120129     DUSP1    25054.85          365 19007.41      834
5 ENSG00000204388    HSPA1B    19085.73          841 16440.69     1197
6 ENSG00000130222   GADD45G    16565.74         1191 13815.91     1592
     d_diff nSD_dev_sex r_diff nSD_rank_sex nSD_bin_rank rank_outlier
1 0.2843081    8.516661    617    15.837301      [12,18)         TRUE
2 0.1392207    4.030301    267     6.853419       [6,12)         TRUE
3 0.1397297    4.046039    285     7.315447       [6,12)         TRUE
4 0.3181621    9.563487    469    12.038402      [12,18)         TRUE
5 0.1608836    4.700155    356     9.137892       [6,12)         TRUE
6 0.1990332    5.879808    401    10.292962       [6,12)         TRUE</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>bias_rank<span class="sc">$</span>sex<span class="sc">$</span>Plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/bias detect rank plot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can change the data point shape using <strong>plot_point_shape</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># shape default = 16</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>bias_rank_shape <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"rank"</span>, <span class="at">nSD_rank =</span> <span class="dv">6</span>, <span class="at">plot_point_shape =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">18</span>))</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(bias_rank_shape<span class="sc">$</span>sample_id<span class="sc">$</span>Plot,bias_rank_shape<span class="sc">$</span>sex<span class="sc">$</span>Plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/shape change-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<ul>
<li><code>threshold = "both"</code>: Detects biased genes based on both deviance change and rank difference, providing a more stringent filtering approach.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>bias_both <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, <span class="at">threshold =</span> <span class="st">"both"</span>,</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">nSD_dev =</span> <span class="dv">7</span>, <span class="at">nSD_rank =</span> <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>bias_both<span class="sc">$</span>sample_id<span class="sc">$</span>Plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/both plot-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bias_both<span class="sc">$</span>sex<span class="sc">$</span>Table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default   dev_sex rank_sex
1 ENSG00000173110     HSPA6    9887.942         2011  7867.831     2074
2 ENSG00000159388      BTG2   18311.285          926 14257.704     1543
3 ENSG00000135625      EGR4   20336.842          705 17851.538      972
4 ENSG00000120738      EGR1   19882.541          752 17444.962     1037
5 ENSG00000120129     DUSP1   25054.848          365 19007.410      834
6 ENSG00000204389    HSPA1A   52523.899           47 41069.968       75
     d_diff nSD_dev_sex r_diff nSD_rank_sex nSD_bin_dev dev_outlier
1 0.2567557    7.664691     63    1.6170988      [7,14)        TRUE
2 0.2843081    8.516661    617   15.8373012      [7,14)        TRUE
3 0.1392207    4.030301    267    6.8534188       [0,7)       FALSE
4 0.1397297    4.046039    285    7.3154471       [0,7)       FALSE
5 0.3181621    9.563487    469   12.0384024      [7,14)        TRUE
6 0.2788882    8.349068     28    0.7187106      [7,14)        TRUE
  nSD_bin_rank rank_outlier
1        [0,6)        FALSE
2      [12,18)         TRUE
3       [6,12)         TRUE
4       [6,12)         TRUE
5      [12,18)         TRUE
6        [0,6)        FALSE</code></pre>
</div>
</div>
<p>We can change the data point color using <strong>plot_pallete</strong>. The color pallete <a href="https://r-graph-gallery.com/38-rcolorbrewers-palettes.html">here</a> can be referenced on since the function uses <code>RColorBrewer</code> to generate colors.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># color default = "YlOrRd"</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>bias_both_color <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"both"</span>, <span class="at">nSD_dev =</span> <span class="dv">7</span>, <span class="at">nSD_rank =</span> <span class="dv">6</span>, <span class="at">plot_palette =</span> <span class="st">"Greens"</span>)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(bias_both_color<span class="sc">$</span>sample_id<span class="sc">$</span>Plot,bias_both_color<span class="sc">$</span>sex<span class="sc">$</span>Plot,<span class="at">nrow =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/color change-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>We can change the text size using <strong>plot_text_size</strong>. We also specify the color palletes for both batch effects at the same time.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># text size default = 3</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>bias_both_color_text <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"both"</span>, <span class="at">nSD_dev =</span> <span class="dv">7</span>, <span class="at">nSD_rank =</span> <span class="dv">6</span>, </span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot_palette =</span> <span class="fu">c</span>(<span class="st">"Blues"</span>,<span class="st">"Greens"</span>), <span class="at">plot_text_size =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>))</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(bias_both_color_text<span class="sc">$</span>sample_id<span class="sc">$</span>Plot,</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>    bias_both_color_text<span class="sc">$</span>sex<span class="sc">$</span>Plot,<span class="at">nrow =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="humanHippocampus2024_sub4_spe_files/figure-html/text change-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section>
<section id="refine-svgs-by-removing-batch-affected-outliers" class="level4">
<h4 class="anchored" data-anchor-id="refine-svgs-by-removing-batch-affected-outliers">Refine SVGs by Removing Batch-Affected Outliers</h4>
<p>Finally, we obtain a refined set of spatially variable genes (SVGs) by removing the identified outliers based on user-defined thresholds for <code>nSD_dev</code> and <code>nSD_rank</code>.</p>
<p>Here, we use the results from bias_both, which applied <code>threshold = "both"</code> to account for both deviance and rank differences, with the batch effect set to sample ID.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>bias_both_df <span class="ot">&lt;-</span> bias_both<span class="sc">$</span>sample_id<span class="sc">$</span>Table</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>svgs_filt <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(svgs_sub4<span class="sc">$</span>gene_id, bias_both_df<span class="sc">$</span>gene_id)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>svgs_sub4_filt <span class="ot">&lt;-</span> svgs_sub4[svgs_sub4<span class="sc">$</span>gene_id <span class="sc">%in%</span> svgs_filt, ]</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(svgs_sub4_filt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2067</code></pre>
</div>
</div>
<p>After obtaining the refined set of SVGs, these genes can be further analyzed using established spatial transcriptomics clustering algorithms to explore tissue layers and spatial organization.</p>
</section>
</section>
<section id="compare-clustering-performances" class="level3">
<h3 class="anchored" data-anchor-id="compare-clustering-performances">Compare clustering performances</h3>
<section id="old-svgs" class="level4">
<h4 class="anchored" data-anchor-id="old-svgs">Old SVGs</h4>
</section>
<section id="refined-svgs" class="level4">
<h4 class="anchored" data-anchor-id="refined-svgs">Refined SVGs</h4>
</section>
<section id="plots-comparison" class="level4">
<h4 class="anchored" data-anchor-id="plots-comparison">Plots comparison</h4>
</section>
</section>
<section id="r-session-information" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="r-session-information"><code>R</code> session information</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Session info</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] humanHippocampus2024_0.99.8 gridExtra_2.3              
 [3] ggspavis_1.12.0             ggplot2_3.5.1              
 [5] BatchSVG_0.99.6             Seurat_5.2.1               
 [7] SeuratObject_5.0.2          sp_2.2-0                   
 [9] PRECAST_1.6.5               gtools_3.9.5               
[11] here_1.0.1                  cowplot_1.1.3              
[13] tibble_3.2.1                dplyr_1.1.4                
[15] tidyr_1.3.1                 SpatialExperiment_1.16.0   
[17] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
[19] Biobase_2.66.0              GenomicRanges_1.58.0       
[21] GenomeInfoDb_1.42.3         IRanges_2.40.1             
[23] S4Vectors_0.44.0            MatrixGenerics_1.18.1      
[25] matrixStats_1.5.0           ExperimentHub_2.14.0       
[27] AnnotationHub_3.14.0        BiocFileCache_2.14.0       
[29] dbplyr_2.5.0                BiocGenerics_0.52.0        

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22        splines_4.4.2           later_1.4.1            
  [4] bitops_1.0-9            filelock_1.0.3          polyclip_1.10-7        
  [7] fastDummies_1.7.5       lifecycle_1.0.4         rstatix_0.7.2          
 [10] rprojroot_2.0.4         globals_0.16.3          lattice_0.22-6         
 [13] MASS_7.3-65             backports_1.5.0         magrittr_2.0.3         
 [16] plotly_4.10.4           rmarkdown_2.29          yaml_2.3.10            
 [19] httpuv_1.6.15           sctransform_0.4.1       ggside_0.3.1           
 [22] spam_2.11-1             spatstat.sparse_3.1-0   reticulate_1.41.0.1    
 [25] pbapply_1.7-2           DBI_1.2.3               RColorBrewer_1.1-3     
 [28] abind_1.4-8             zlibbioc_1.52.0         Rtsne_0.17             
 [31] purrr_1.0.4             RCurl_1.98-1.17         rappdirs_0.3.3         
 [34] GenomeInfoDbData_1.2.13 ggrepel_0.9.6           scry_1.18.0            
 [37] irlba_2.3.5.1           listenv_0.9.1           spatstat.utils_3.1-3   
 [40] goftest_1.2-3           RSpectra_0.16-2         spatstat.random_3.3-3  
 [43] fitdistrplus_1.2-2      parallelly_1.43.0       codetools_0.2-20       
 [46] DelayedArray_0.32.0     scuttle_1.16.0          tidyselect_1.2.1       
 [49] UCSC.utils_1.2.0        farver_2.1.2            viridis_0.6.5          
 [52] ScaledMatrix_1.14.0     spatstat.explore_3.4-2  jsonlite_1.9.1         
 [55] BiocNeighbors_2.0.1     Formula_1.2-5           progressr_0.15.1       
 [58] scater_1.34.1           ggridges_0.5.6          survival_3.8-3         
 [61] tools_4.4.2             ica_1.0-3               Rcpp_1.0.14            
 [64] glue_1.8.0              SparseArray_1.6.2       xfun_0.51              
 [67] ggthemes_5.1.0          withr_3.0.2             BiocManager_1.30.25    
 [70] fastmap_1.2.0           rsvd_1.0.5              digest_0.6.37          
 [73] R6_2.6.1                mime_0.13               colorspace_2.1-1       
 [76] scattermore_1.2         tensor_1.5              spatstat.data_3.1-6    
 [79] RSQLite_2.3.9           generics_0.1.3          data.table_1.17.0      
 [82] httr_1.4.7              htmlwidgets_1.6.4       S4Arrays_1.6.0         
 [85] uwot_0.2.3              pkgconfig_2.0.3         gtable_0.3.6           
 [88] blob_1.2.4              lmtest_0.9-40           XVector_0.46.0         
 [91] htmltools_0.5.8.1       carData_3.0-5           dotCall64_1.2          
 [94] scales_1.3.0            png_0.1-8               spatstat.univar_3.1-2  
 [97] knitr_1.50              rstudioapi_0.17.1       reshape2_1.4.4         
[100] rjson_0.2.23            nlme_3.1-167            curl_6.2.2             
[103] cachem_1.1.0            zoo_1.8-13              stringr_1.5.1          
[106] BiocVersion_3.20.0      KernSmooth_2.23-26      vipor_0.4.7            
[109] miniUI_0.1.1.1          GiRaF_1.0.1             AnnotationDbi_1.68.0   
[112] pillar_1.10.1           grid_4.4.2              vctrs_0.6.5            
[115] RANN_2.6.2              ggpubr_0.6.0            promises_1.3.2         
[118] car_3.1-3               BiocSingular_1.22.0     beachmat_2.22.0        
[121] DR.SC_3.4               xtable_1.8-4            cluster_2.1.8.1        
[124] beeswarm_0.4.0          evaluate_1.0.3          magick_2.8.6           
[127] cli_3.6.4               compiler_4.4.2          rlang_1.1.5            
[130] crayon_1.5.3            ggsignif_0.6.4          future.apply_1.11.3    
[133] labeling_0.4.3          mclust_6.1.1            ggbeeswarm_0.7.2       
[136] plyr_1.8.9              stringi_1.8.4           BiocParallel_1.40.0    
[139] viridisLite_0.4.2       deldir_2.0-4            munsell_0.5.1          
[142] Biostrings_2.74.1       lazyeval_0.2.2          spatstat.geom_3.3-6    
[145] CompQuadForm_1.4.3      Matrix_1.7-3            RcppHNSW_0.6.0         
[148] patchwork_1.3.0         bit64_4.6.0-1           future_1.34.0          
[151] KEGGREST_1.46.0         shiny_1.10.0            ROCR_1.0-11            
[154] broom_1.0.7             igraph_2.1.4            memoise_2.0.1          
[157] bit_4.6.0              </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>