<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>BatchSVG_analysis - spatialLIBD_spe</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">BatchSVG_analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./featureSelection.html" rel="" target="">
 <span class="menu-text">Why</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./dataDrivenThreshold.html" rel="" target="">
 <span class="menu-text">How</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./humanHippocampus2024_sub4_spe.html" rel="" target="">
 <span class="menu-text">analysis_1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./spatialLIBD_spe.html" rel="" target="" aria-current="page">
 <span class="menu-text">analysis_2</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#installation" id="toc-installation" class="nav-link" data-scroll-target="#installation">Installation</a></li>
  <li><a href="#biased-feature-identification" id="toc-biased-feature-identification" class="nav-link" data-scroll-target="#biased-feature-identification">Biased Feature Identification</a></li>
  <li><a href="#r-session-information" id="toc-r-session-information" class="nav-link" data-scroll-target="#r-session-information"><code>R</code> session information</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">spatialLIBD_spe</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="introduction" class="level3">
<h3 class="anchored" data-anchor-id="introduction">Introduction</h3>
<p><code>BatchSVG</code> is the R/Bioconductor package for spatial transcriptomics data quality control (QC). As the feature-based QC method, the package provides functions to identify the biased features associated with the batch effect(s) (e.g.&nbsp;sample, slide, and sex) in spatially variable genes (SVGs) using binomial deviance model, aiming to develop the downstream clustering performances and remove the technical noises caused by batch effects. The package works with <a href="https://github.com/drighelli/SpatialExperiment">SpatialExperiment</a> objects.</p>
</section>
<section id="installation" class="level3">
<h3 class="anchored" data-anchor-id="installation">Installation</h3>
<p>(After accepted in <a href="https://bioconductor.org/">Bioconductor</a>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>)) {</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"BatchSVG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Install the development version from <a href="https://christinehou11.github.io/BatchSVG">GitHub</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install</span>(<span class="st">"christinehou11/BatchSVG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="biased-feature-identification" class="level3">
<h3 class="anchored" data-anchor-id="biased-feature-identification">Biased Feature Identification</h3>
<p>In this section, we will include the standard workflow for using <code>BatchSVG</code> to show how the method help to detect and visualize the biased features in SVGs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BatchSVG)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(spatialLIBD)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="data" class="level4">
<h4 class="anchored" data-anchor-id="data">Data</h4>
<p>We will use the`spatially-resolved transcriptomics (SRT) dataset from the <a href="https://research.libd.org/spatialLIBD/">spatialLIBD</a> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>spatialLIBD_spe <span class="ot">&lt;-</span> <span class="fu">fetch_data</span>(<span class="at">type =</span> <span class="st">"spe"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>spatialLIBD_spe</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SpatialExperiment 
dim: 33538 47681 
metadata(0):
assays(2): counts logcounts
rownames(33538): ENSG00000243485 ENSG00000237613 ... ENSG00000277475
  ENSG00000268674
rowData names(9): source type ... gene_search is_top_hvg
colnames(47681): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...
  TTGTTTCCATACAACT-1 TTGTTTGTGTAAATTC-1
colData names(69): sample_id Cluster ... array_row array_col
reducedDimNames(6): PCA TSNE_perplexity50 ... TSNE_perplexity80
  UMAP_neighbors15
mainExpName: NULL
altExpNames(0):
spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres
imgData names(4): sample_id image_id data scaleFactor</code></pre>
</div>
</div>
<p>We will use the spatially variable genes set generated. The result is generated from <a href="(https://www.nature.com/articles/s41467-023-39748-z)">nnSVG</a> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>libd_nnsvgs <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="fu">here</span>(<span class="st">"data/libd-all_nnSVG_p-05-features-df.csv"</span>),</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">row.names =</span> <span class="dv">1</span>, <span class="at">check.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="perform-feature-selection-using-featureselect" class="level4">
<h4 class="anchored" data-anchor-id="perform-feature-selection-using-featureselect">Perform Feature Selection using <code>featureSelect()</code></h4>
<p>We will perform feature selection on the spatial transcriptomics data (<em>input</em>) using a predefined set of spatially variable genes (<em>VGs</em>). Specifically, we will compute the number of standard deviations for the relative change in deviance (<strong>nSD_dev_{batch effect}</strong>) and rank difference (<strong>nSD_rank_{batch effect}</strong>) before and after adjusting for batch effects.</p>
<p>The <code>featureSelect()</code> function enables feature selection while accounting for multiple batch effects. It returns a <strong>list</strong> of data frames, where each batch effect is associated with a corresponding data frame containing key results, including:</p>
<ul>
<li><p>Relative change in deviance before and after batch effect adjustment</p></li>
<li><p>Rank differences between the batch-corrected and uncorrected results</p></li>
<li><p>Number of standard deviations (nSD) for both relative change in deviance and rank difference</p></li>
</ul>
<p>We will use the example of applying <code>featureSelect()</code> to the sample dataset while adjusting for the batch effect of <em>subject</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>list_batch_df <span class="ot">&lt;-</span> <span class="fu">featureSelect</span>(<span class="at">input =</span> spatialLIBD_spe, </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">batch_effect =</span> <span class="st">"subject"</span>, <span class="at">VGs =</span> libd_nnsvgs<span class="sc">$</span>gene_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Running feature selection without batch...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Batch Effect: subject</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Running feature selection without batch...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating deviance and rank difference...</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(list_batch_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "list"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(list_batch_df<span class="sc">$</span>subject)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default dev_subject rank_subject
1 ENSG00000187608     ISG15    43591.43         1151    43304.39         1157
2 ENSG00000131584     ACAP3    39115.97         1498    38826.05         1509
3 ENSG00000242485    MRPL20    47074.23          920    46912.53          918
4 ENSG00000160075     SSU72    47815.13          862    47742.10          852
5 ENSG00000078369      GNB1    54464.75          477    54385.81          462
6 ENSG00000187730     GABRD    51521.13          646    51099.35          648
       d_diff nSD_dev_subject r_diff nSD_rank_subject
1 0.006628419     -0.04505973      6       0.15937488
2 0.007467317     -0.03697398     11       0.29218728
3 0.003446929     -0.07572463     -2      -0.05312496
4 0.001529688     -0.09420402    -10      -0.26562480
5 0.001451448     -0.09495814    -15      -0.39843720
6 0.008254157     -0.02939000      2       0.05312496</code></pre>
</div>
</div>
</section>
<section id="visualize-svg-selection-using-svg_nsd-for-batch-effects" class="level4">
<h4 class="anchored" data-anchor-id="visualize-svg-selection-using-svg_nsd-for-batch-effects">Visualize SVG Selection Using <code>svg_nSD</code> for Batch Effects</h4>
<p>The <code>svg_nSD()</code> function generates visualizations to assess batch effects in spatially variable genes (SVGs). It produces bar charts showing the distribution of SVGs based on relative change in deviance and rank difference, with colors representing different nSD intervals. Additionally, scatter plots compare deviance and rank values with and without batch effects.</p>
<p>By interpreting these plots, we can determine appropriate nSD thresholds for filtering biased features. The left panels illustrate the distribution of SVGs in terms of deviance and rank difference, while the right panels compare values before and after accounting for batch effects.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">svg_nSD</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">sd_interval_dev =</span> <span class="dv">3</span>, <span class="at">sd_interval_rank =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Figure 1. Visualizations of nSD_dev and nSD_rank threshold selection</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>plots<span class="sc">$</span>subject</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="spatialLIBD_spe_files/figure-html/figure 1-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section>
<section id="identify-biased-genes-using-biasdetect" class="level4">
<h4 class="anchored" data-anchor-id="identify-biased-genes-using-biasdetect">Identify Biased Genes Using <code>biasDetect()</code></h4>
<p>The function <code>biasDetect()</code> is designed to identify and filter out biased genes across different batch effects. Using threshold values selected from the visualization results generated by <code>svg_nSD()</code>, this function systematically detects outliers that exceed a specified number of standard deviation (nSD) threshold in either relative deviance change, rank difference, or both.</p>
<p>The function outputs visualizations comparing deviance and rank values with and without batch effects. Genes with high deviations, highlighted in color, are identified as potentially biased and can be excluded based on the selected nSD thresholds.</p>
<p>The function offers flexibility in customizing the plot aesthetics, allowing users to adjust the data point size (<strong>plot_point_size</strong>), shape (<strong>plot_point_shape</strong>), annotated text size (<strong>plot_text_size</strong>), and data point color palette (<strong>plot_palette</strong>). Default values are provided for these parameters if not specified. Users should refer to <a href="https://ggplot2.tidyverse.org/index.html">ggplot2</a> aesthetic guidelines to ensure appropriate values are assigned for each parameter.</p>
<p>We will use <code>nSD_dev = 3</code> and <code>nSD_rank = 3</code> as the example. The user should adjust the value based on their dataset features.</p>
<p><strong>Usage of Different Threshold Options</strong></p>
<ul>
<li><code>threshold = "dev"</code>: Filters biased genes based only on the relative change in deviance. Genes with deviance changes exceeding the specified <code>nSD_dev</code> threshold are identified as batch-affected and can be removed.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>bias_dev <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"dev"</span>, <span class="at">nSD_dev =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Table 1. Outlier Genes defined by nSD_dev only</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bias_dev<span class="sc">$</span>subject<span class="sc">$</span>Table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default dev_subject rank_subject
1 ENSG00000255823  MTRNR2L8    286927.9            5    67738.95           77
2 ENSG00000087250       MT3    115108.2           21    86093.50           31
3 ENSG00000256618  MTRNR2L1    167381.9           14    40299.65         1368
4 ENSG00000198840    MT-ND3    170192.2           12   120647.03           17
     d_diff nSD_dev_subject r_diff nSD_rank_subject nSD_bin_dev dev_outlier
1 3.2357886       31.079310     72        1.9124986     [30,33]        TRUE
2 0.3370142        3.139375     10        0.2656248       [3,6)        TRUE
3 3.1534316       30.285509   1354       35.9655982     [30,33]        TRUE
4 0.4106625        3.849237      5        0.1328124       [3,6)        TRUE</code></pre>
</div>
</div>
<p>We can change the data point size using <strong>plot_point_size</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># size default = 3</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>bias_dev_size <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"dev"</span>, <span class="at">nSD_dev =</span> <span class="dv">3</span>, <span class="at">plot_point_size =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Figure 2. Customize point size</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(bias_dev<span class="sc">$</span>subject<span class="sc">$</span>Plot, bias_dev_size<span class="sc">$</span>subject<span class="sc">$</span>Plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="spatialLIBD_spe_files/figure-html/figure 2-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<ul>
<li><code>threshold = "rank"</code>: Identifies biased genes based solely on rank difference. Genes with rank shifts exceeding <code>nSD_rank</code> are considered biased.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>bias_rank <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"rank"</span>, <span class="at">nSD_rank =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Table 2. Outlier Genes defined by nSD_rank only</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bias_rank<span class="sc">$</span>subject<span class="sc">$</span>Table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default dev_subject rank_subject
1 ENSG00000184144     CNTN2    44388.43         1101    42016.77         1258
2 ENSG00000136541      ERMN    48106.50          844    44879.30         1036
3 ENSG00000013297    CLDN11    53385.84          539    49776.21          732
4 ENSG00000164124   TMEM144    39407.07         1471    37587.88         1605
5 ENSG00000204655       MOG    37293.40         1648    35613.38         1762
6 ENSG00000091129     NRCAM    51388.42          655    49006.18          776
      d_diff nSD_dev_subject r_diff nSD_rank_subject nSD_bin_rank rank_outlier
1 0.05644564       0.4351052    157         4.170309        [3,6)         TRUE
2 0.07190853       0.5841448    192         5.099996        [3,6)         TRUE
3 0.07251725       0.5900120    193         5.126559        [3,6)         TRUE
4 0.04839817       0.3575395    134         3.559372        [3,6)         TRUE
5 0.04717397       0.3457399    114         3.028123        [3,6)         TRUE
6 0.04861096       0.3595904    121         3.214060        [3,6)         TRUE</code></pre>
</div>
</div>
<p>We can change the data point shape using <strong>plot_point_shape</strong>.</p>
<p><em>Figure 3. Customize point shape</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># shape default = 16</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>bias_rank_shape <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"rank"</span>, <span class="at">nSD_rank =</span> <span class="dv">3</span>, <span class="at">plot_point_shape =</span> <span class="dv">2</span>)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(bias_rank<span class="sc">$</span>subject<span class="sc">$</span>Plot, bias_rank_shape<span class="sc">$</span>subject<span class="sc">$</span>Plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="spatialLIBD_spe_files/figure-html/figure 3-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<ul>
<li><code>threshold = "both"</code>: Detects biased genes based on both deviance change and rank difference, providing a more stringent filtering approach.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>bias_both <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, <span class="at">threshold =</span> <span class="st">"both"</span>,</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">nSD_dev =</span> <span class="dv">3</span>, <span class="at">nSD_rank =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Table 3. Outlier Genes defined by nSD_dev and nSD_rank</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bias_both<span class="sc">$</span>subject<span class="sc">$</span>Table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default dev_subject rank_subject
1 ENSG00000184144     CNTN2    44388.43         1101    42016.77         1258
2 ENSG00000136541      ERMN    48106.50          844    44879.30         1036
3 ENSG00000013297    CLDN11    53385.84          539    49776.21          732
4 ENSG00000164124   TMEM144    39407.07         1471    37587.88         1605
5 ENSG00000204655       MOG    37293.40         1648    35613.38         1762
6 ENSG00000091129     NRCAM    51388.42          655    49006.18          776
      d_diff nSD_dev_subject r_diff nSD_rank_subject nSD_bin_dev dev_outlier
1 0.05644564       0.4351052    157         4.170309       [0,3)       FALSE
2 0.07190853       0.5841448    192         5.099996       [0,3)       FALSE
3 0.07251725       0.5900120    193         5.126559       [0,3)       FALSE
4 0.04839817       0.3575395    134         3.559372       [0,3)       FALSE
5 0.04717397       0.3457399    114         3.028123       [0,3)       FALSE
6 0.04861096       0.3595904    121         3.214060       [0,3)       FALSE
  nSD_bin_rank rank_outlier
1        [3,6)         TRUE
2        [3,6)         TRUE
3        [3,6)         TRUE
4        [3,6)         TRUE
5        [3,6)         TRUE
6        [3,6)         TRUE</code></pre>
</div>
</div>
<p>We can change the data point color using <strong>plot_palette</strong>. The color palette <a href="https://r-graph-gallery.com/38-rcolorbrewers-palettes.html">here</a> can be referenced on since the function uses <code>RColorBrewer</code> to generate colors.</p>
<p><em>Figure 4. Customize the palette color</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># color default = "YlOrRd"</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>bias_both_color <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"both"</span>, <span class="at">nSD_dev =</span> <span class="dv">3</span>, <span class="at">nSD_rank =</span> <span class="dv">3</span>, <span class="at">plot_palette =</span> <span class="st">"Greens"</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(bias_both<span class="sc">$</span>subject<span class="sc">$</span>Plot, bias_both_color<span class="sc">$</span>subject<span class="sc">$</span>Plot,<span class="at">nrow =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="spatialLIBD_spe_files/figure-html/figure 4-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>We can change the text size using <strong>plot_text_size</strong>. We also specify the color palettes for both batch effects at the same time.</p>
<p><em>Figure 5. Customize text size and color palette</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># text size default = 3</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>bias_both_color_text <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="st">"both"</span>, <span class="at">nSD_dev =</span> <span class="dv">3</span>, <span class="at">nSD_rank =</span> <span class="dv">3</span>, </span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot_palette =</span> <span class="fu">c</span>(<span class="st">"Blues"</span>), <span class="at">plot_text_size =</span> <span class="dv">4</span>)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(bias_both<span class="sc">$</span>subject<span class="sc">$</span>Plot, bias_both_color_text<span class="sc">$</span>subject<span class="sc">$</span>Plot,<span class="at">nrow =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="spatialLIBD_spe_files/figure-html/figure 5-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section>
<section id="refine-svgs-by-removing-batch-affected-outliers" class="level4">
<h4 class="anchored" data-anchor-id="refine-svgs-by-removing-batch-affected-outliers">Refine SVGs by Removing Batch-Affected Outliers</h4>
<p>Finally, we obtain a refined set of spatially variable genes (SVGs) by removing the identified outliers based on user-defined thresholds for <code>nSD_dev</code> and <code>nSD_rank</code>.</p>
<p>Here, we use the results from bias_both, which applied <code>threshold = "both"</code> to account for both deviance and rank differences, with the batch effect set to sample ID.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>bias_both_df <span class="ot">&lt;-</span> bias_both<span class="sc">$</span>subject<span class="sc">$</span>Table</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>svgs_filt <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(libd_nnsvgs<span class="sc">$</span>gene_id, bias_both_df<span class="sc">$</span>gene_id)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>svgs_filt_spe <span class="ot">&lt;-</span> libd_nnsvgs[libd_nnsvgs<span class="sc">$</span>gene_id <span class="sc">%in%</span> svgs_filt, ]</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(svgs_filt_spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1951</code></pre>
</div>
</div>
<p>After obtaining the refined set of SVGs, these genes can be further analyzed using established spatial transcriptomics clustering algorithms to explore tissue layers and spatial organization.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># </span><span class="al">TBD</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="r-session-information" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="r-session-information"><code>R</code> session information</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Session info</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] here_1.0.1                  cowplot_1.1.3              
 [3] spatialLIBD_1.18.0          SpatialExperiment_1.16.0   
 [5] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
 [7] Biobase_2.66.0              GenomicRanges_1.58.0       
 [9] GenomeInfoDb_1.42.3         IRanges_2.40.1             
[11] S4Vectors_0.44.0            BiocGenerics_0.52.0        
[13] MatrixGenerics_1.18.1       matrixStats_1.5.0          
[15] BatchSVG_0.99.6            

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3       rstudioapi_0.17.1        jsonlite_1.9.1          
  [4] magrittr_2.0.3           ggbeeswarm_0.7.2         magick_2.8.5            
  [7] farver_2.1.2             rmarkdown_2.29           BiocIO_1.16.0           
 [10] fields_16.3              zlibbioc_1.52.0          vctrs_0.6.5             
 [13] Rsamtools_2.22.0         memoise_2.0.1            config_0.3.2            
 [16] paletteer_1.6.0          RCurl_1.98-1.16          benchmarkme_1.0.8       
 [19] htmltools_0.5.8.1        S4Arrays_1.6.0           AnnotationHub_3.14.0    
 [22] curl_6.2.1               BiocNeighbors_2.0.1      SparseArray_1.6.2       
 [25] sass_0.4.9               bslib_0.9.0              htmlwidgets_1.6.4       
 [28] plotly_4.10.4            cachem_1.1.0             GenomicAlignments_1.42.0
 [31] mime_0.12                lifecycle_1.0.4          iterators_1.0.14        
 [34] pkgconfig_2.0.3          rsvd_1.0.5               Matrix_1.7-2            
 [37] R6_2.6.1                 fastmap_1.2.0            GenomeInfoDbData_1.2.13 
 [40] shiny_1.10.0             digest_0.6.37            colorspace_2.1-1        
 [43] rematch2_2.1.2           AnnotationDbi_1.68.0     rprojroot_2.0.4         
 [46] scater_1.34.0            irlba_2.3.5.1            ExperimentHub_2.14.0    
 [49] RSQLite_2.3.9            beachmat_2.22.0          labeling_0.4.3          
 [52] filelock_1.0.3           httr_1.4.7               abind_1.4-8             
 [55] compiler_4.4.2           withr_3.0.2              bit64_4.6.0-1           
 [58] doParallel_1.0.17        attempt_0.3.1            BiocParallel_1.40.0     
 [61] viridis_0.6.5            DBI_1.2.3                maps_3.4.2.1            
 [64] sessioninfo_1.2.3        rappdirs_0.3.3           DelayedArray_0.32.0     
 [67] rjson_0.2.23             tools_4.4.2              vipor_0.4.7             
 [70] beeswarm_0.4.0           httpuv_1.6.15            glue_1.8.0              
 [73] restfulr_0.0.15          promises_1.3.2           grid_4.4.2              
 [76] generics_0.1.3           gtable_0.3.6             tidyr_1.3.1             
 [79] data.table_1.17.0        BiocSingular_1.22.0      ScaledMatrix_1.14.0     
 [82] XVector_0.46.0           stringr_1.5.1            ggrepel_0.9.6           
 [85] BiocVersion_3.20.0       foreach_1.5.2            pillar_1.10.1           
 [88] spam_2.11-1              limma_3.62.2             later_1.4.1             
 [91] benchmarkmeData_1.0.4    dplyr_1.1.4              BiocFileCache_2.14.0    
 [94] lattice_0.22-6           rtracklayer_1.66.0       bit_4.6.0               
 [97] tidyselect_1.2.1         locfit_1.5-9.11          scuttle_1.16.0          
[100] Biostrings_2.74.1        knitr_1.49               gridExtra_2.3           
[103] edgeR_4.4.2              xfun_0.51                statmod_1.5.0           
[106] DT_0.33                  stringi_1.8.4            UCSC.utils_1.2.0        
[109] lazyeval_0.2.2           yaml_2.3.10              shinyWidgets_0.9.0      
[112] evaluate_1.0.3           codetools_0.2-20         tibble_3.2.1            
[115] BiocManager_1.30.25      cli_3.6.4                xtable_1.8-4            
[118] jquerylib_0.1.4          munsell_0.5.1            golem_0.5.1             
[121] Rcpp_1.0.14              dbplyr_2.5.0             scry_1.18.0             
[124] png_0.1-8                XML_3.99-0.18            parallel_4.4.2          
[127] ggplot2_3.5.1            blob_1.2.4               dotCall64_1.2           
[130] bitops_1.0-9             viridisLite_0.4.2        scales_1.3.0            
[133] purrr_1.0.4              crayon_1.5.3             rlang_1.1.5             
[136] KEGGREST_1.46.0         </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>