<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>BatchSVG_analysis - spatialLIBD Dataset</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">BatchSVG_analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./featureSelection.html" rel="" target="">
 <span class="menu-text">Why</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./dataDrivenThreshold.html" rel="" target="">
 <span class="menu-text">How</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./humanHippocampus2024_sub4_spe.html" rel="" target="">
 <span class="menu-text">analysis_1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./spatialLIBD_spe.html" rel="" target="" aria-current="page">
 <span class="menu-text">analysis_2</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#before-batchsvg-implementation" id="toc-before-batchsvg-implementation" class="nav-link" data-scroll-target="#before-batchsvg-implementation">Before <code>BatchSVG</code> implementation</a></li>
  <li><a href="#batchsvg-implementation" id="toc-batchsvg-implementation" class="nav-link" data-scroll-target="#batchsvg-implementation"><code>BatchSVG</code> implementation</a></li>
  <li><a href="#after-batchsvg-implementation" id="toc-after-batchsvg-implementation" class="nav-link" data-scroll-target="#after-batchsvg-implementation">After <code>BatchSVG</code> implementation</a></li>
  <li><a href="#r-session-information" id="toc-r-session-information" class="nav-link" data-scroll-target="#r-session-information"><code>R</code> session information</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">spatialLIBD Dataset</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="introduction" class="level3">
<h3 class="anchored" data-anchor-id="introduction">Introduction</h3>
<p>In this section, we will include biased feature identification analysis using our developed method to show how <code>BatchSVG</code> helps to detect and visualize the biased features in SVGs. We will use the `spatially-resolved transcriptomics (SRT) dataset from the <a href="https://research.libd.org/spatialLIBD/">spatialLIBD</a> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(spatialLIBD)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(nnSVG)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(PRECAST)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BatchSVG)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggspavis)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gridExtra)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>spatialLIBD</code> dataset was collected from the brain region of dorsolateral prefontal cortex (dlPFC) and was also sequenced with 10X Visium. This dataset has 3 subjects and 4 samples from each subject. Because there was no slide metadata available, we investigated whether <code>batch = subject</code> influenced the top features.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load example dataset from spatialLIBD package</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># see '?spatialLIBD' for more details</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>spatialLIBD_spe <span class="ot">&lt;-</span> <span class="fu">fetch_data</span>(<span class="at">type =</span> <span class="st">"spe"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spatialLIBD_spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 33538 47681</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">colData</span>(spatialLIBD_spe)[,<span class="fu">c</span>(<span class="st">"sample_id"</span>,<span class="st">"subject"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         subject
sample_id Br5292 Br5595 Br8100
   151507   4226      0      0
   151508   4384      0      0
   151509   4789      0      0
   151510   4634      0      0
   151669      0   3661      0
   151670      0   3498      0
   151671      0   4110      0
   151672      0   4015      0
   151673      0      0   3639
   151674      0      0   3673
   151675      0      0   3592
   151676      0      0   3460</code></pre>
</div>
</div>
</section>
<section id="before-batchsvg-implementation" class="level3">
<h3 class="anchored" data-anchor-id="before-batchsvg-implementation">Before <code>BatchSVG</code> implementation</h3>
<p>We will use the spatially variable genes set generated. The result is generated from <a href="https://www.nature.com/articles/s41467-023-39748-z">nnSVG</a> package.</p>
<section id="preprocessing" class="level4">
<h4 class="anchored" data-anchor-id="preprocessing">Preprocessing</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># see '?nnSVG' for more details</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># keep only spots over tissue</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>spatialLIBD_spe_1 <span class="ot">&lt;-</span> spatialLIBD_spe[, <span class="fu">colData</span>(spatialLIBD_spe)<span class="sc">$</span>in_tissue <span class="sc">==</span> <span class="dv">1</span>]</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spatialLIBD_spe_1) <span class="co"># 33538 47681</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 33538 47681</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter any new zeros created</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remove genes with zero expression</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>ix_zero_genes <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">counts</span>(spatialLIBD_spe_1)) <span class="sc">==</span> <span class="dv">0</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ix_zero_genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ix_zero_genes
FALSE  TRUE 
25615  7923 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">sum</span>(ix_zero_genes) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  spatialLIBD_spe_2 <span class="ot">&lt;-</span> spatialLIBD_spe_1[<span class="sc">!</span>ix_zero_genes, ]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spatialLIBD_spe_2) <span class="co"># 25615 47681</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 25615 47681</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove spots with zero expression</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>ix_zero_spots <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">counts</span>(spatialLIBD_spe_2)) <span class="sc">==</span> <span class="dv">0</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ix_zero_spots)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ix_zero_spots
FALSE 
47681 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">sum</span>(ix_zero_spots) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  spatialLIBD_spe_3 <span class="ot">&lt;-</span> spatialLIBD_spe_2[, <span class="sc">!</span>ix_zero_spots]</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ix_zero_spots</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co"># FALSE </span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 47681</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>spatialLIBD_spe_filtered <span class="ot">&lt;-</span> spatialLIBD_spe_2</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(spatialLIBD_spe_filtered)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 25615 47681</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>sample_ids <span class="ot">&lt;-</span> <span class="fu">levels</span>(<span class="fu">factor</span>(<span class="fu">colData</span>(spatialLIBD_spe_filtered)<span class="sc">$</span>sample_id))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>sample_ids</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "151507" "151508" "151509" "151510" "151669" "151670" "151671" "151672"
 [9] "151673" "151674" "151675" "151676"</code></pre>
</div>
</div>
</section>
<section id="run-nnsvg-per-sample-and-store-lists-of-top-svgs" class="level4">
<h4 class="anchored" data-anchor-id="run-nnsvg-per-sample-and-store-lists-of-top-svgs">Run <code>nnSVG</code> per sample and store lists of top SVGs</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we offer a pseducode example demonstrating how to run `nnSVG` for multiple samples.</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>res_list <span class="ot">&lt;-</span> <span class="fu">as.list</span>(<span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">length</span>(sample_ids)))</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res_list) <span class="ot">&lt;-</span> sample_ids</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (s <span class="cf">in</span> <span class="fu">seq_along</span>(sample_ids[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])) {</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select sample</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>  ix <span class="ot">&lt;-</span> <span class="fu">colData</span>(spatialLIBD_spe_filtered)<span class="sc">$</span>sample_id <span class="sc">==</span> sample_ids[s]</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>  spe_sub <span class="ot">&lt;-</span> spatialLIBD_spe_filtered[, ix]</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dim</span>(spe_sub)</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># run nnSVG filtering for mitochondrial genes and low-expressed genes</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>  spe_sub <span class="ot">&lt;-</span> <span class="fu">filter_genes</span>(</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>    spe_sub, </span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">filter_genes_ncounts =</span> <span class="dv">3</span>, </span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">filter_genes_pcspots =</span> <span class="fl">0.5</span>, </span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">filter_mito =</span> <span class="cn">TRUE</span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remove any zeros introduced by filtering</span></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>  ix_zeros <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">counts</span>(spe_sub)) <span class="sc">==</span> <span class="dv">0</span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">sum</span>(ix_zeros) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>    spe_sub <span class="ot">&lt;-</span> spe_sub[, <span class="sc">!</span>ix_zeros]</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dim</span>(spe_sub)</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># re-calculate logcounts after filtering</span></span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>  spe_sub <span class="ot">&lt;-</span> <span class="fu">computeLibraryFactors</span>(spe_sub)</span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>  spe_sub <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe_sub)</span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a>  <span class="co"># run nnSVG</span></span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a>  spe_sub <span class="ot">&lt;-</span> <span class="fu">nnSVG</span>(spe_sub)</span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a>  <span class="co"># store results for this sample</span></span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true" tabindex="-1"></a>  res_list[[s]] <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe_sub)</span>
<span id="cb18-38"><a href="#cb18-38" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, combine the results across multiple samples by averaging the ranks of the SVGs following the steps showing in the <code>nnSVG</code> <a href="https://github.com/lmweber/nnSVG/blob/4e6427e58e6c77119d1d74237c540abc6bfa8713/vignettes/nnSVG.Rmd#L476">vignette</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># number of genes that passed filtering (and subsampling) for each sample</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>(res_list, nrow)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co"># match results from each sample and store in matching rows</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>res_ranks_LIBD <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="fu">nrow</span>(spatialLIBD_spe_filtered), <span class="at">ncol =</span> <span class="fu">length</span>(sample_ids))</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(res_ranks_LIBD) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(spatialLIBD_spe_filtered)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(res_ranks_LIBD) <span class="ot">&lt;-</span> sample_ids</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (s <span class="cf">in</span> <span class="fu">seq_along</span>(sample_ids)) {</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stopifnot</span>(<span class="fu">colnames</span>(res_ranks_LIBD)[s] <span class="sc">==</span> sample_ids[s])</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stopifnot</span>(<span class="fu">colnames</span>(res_ranks_LIBD)[s] <span class="sc">==</span> <span class="fu">names</span>(res_list)[s])</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>  rownames_s <span class="ot">&lt;-</span> <span class="fu">rownames</span>(res_list[[s]])</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>  res_ranks_LIBD[rownames_s, s] <span class="ot">&lt;-</span> res_list[[s]][, <span class="st">"rank"</span>]</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="co"># remove genes that were filtered out in all samples</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>ix_allna <span class="ot">&lt;-</span> <span class="fu">apply</span>(res_ranks_LIBD, <span class="dv">1</span>, <span class="cf">function</span>(r) <span class="fu">all</span>(<span class="fu">is.na</span>(r)))</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>res_ranks_LIBD <span class="ot">&lt;-</span> res_ranks_LIBD[<span class="sc">!</span>ix_allna, ]</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_ranks_LIBD)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1967</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>                        source type         gene_id gene_version gene_name
ENSG00000187608 ensembl_havana gene ENSG00000187608            8     ISG15
ENSG00000131584 ensembl_havana gene ENSG00000131584           18     ACAP3
ENSG00000242485 ensembl_havana gene ENSG00000242485            5    MRPL20
ENSG00000160075 ensembl_havana gene ENSG00000160075           11     SSU72
ENSG00000078369 ensembl_havana gene ENSG00000078369           17      GNB1
ENSG00000187730 ensembl_havana gene ENSG00000187730            8     GABRD
                   gene_source   gene_biotype             gene_search
ENSG00000187608 ensembl_havana protein_coding  ISG15; ENSG00000187608
ENSG00000131584 ensembl_havana protein_coding  ACAP3; ENSG00000131584
ENSG00000242485 ensembl_havana protein_coding MRPL20; ENSG00000242485
ENSG00000160075 ensembl_havana protein_coding  SSU72; ENSG00000160075
ENSG00000078369 ensembl_havana protein_coding   GNB1; ENSG00000078369
ENSG00000187730 ensembl_havana protein_coding  GABRD; ENSG00000187730
                is_top_hvg    sigma.sq    tau.sq          phi    loglik runtime
ENSG00000187608       TRUE 0.014755187 0.3356806 9.119204e+00 -42272.79  23.241
ENSG00000131584      FALSE 0.001055874 0.2868728 6.900461e-02 -37864.96  20.942
ENSG00000242485      FALSE 0.001656123 0.4411004 1.497840e-02 -48094.96  20.343
ENSG00000160075      FALSE 0.003069247 0.4646316 1.716212e-03 -49391.53  19.676
ENSG00000078369      FALSE 0.007099322 0.7113184 3.456360e-07 -59584.20  23.359
ENSG00000187730      FALSE 0.007227928 0.4965150 1.795660e-04 -51122.73  20.104
                     mean       var      spcov     prop_sv loglik_lm    LR_stat
ENSG00000187608 0.3141369 0.3505163 0.38668154 0.042105251 -42543.76 541.936437
ENSG00000131584 0.2860178 0.2879378 0.11360906 0.003667137 -37868.30   6.688740
ENSG00000242485 0.4788772 0.4427726 0.08498106 0.003740482 -48098.53   7.151903
ENSG00000160075 0.5193062 0.4677126 0.10668230 0.006562414 -49401.30  19.526820
ENSG00000078369 1.0635173 0.7184340 0.07922530 0.009881886 -59605.57  42.730757
ENSG00000187730 0.5727988 0.5037425 0.14842423 0.014348446 -51165.59  85.727756
                rank         pval         padj
ENSG00000187608  167 0.000000e+00 0.000000e+00
ENSG00000131584 1953 3.528243e-02 4.669999e-02
ENSG00000242485 1910 2.798878e-02 3.788010e-02
ENSG00000160075 1282 5.751816e-05 1.159785e-04
ENSG00000078369  839 5.261797e-10 1.621185e-09
ENSG00000187730  548 0.000000e+00 0.000000e+00</code></pre>
</div>
</div>
</section>
</section>
<section id="batchsvg-implementation" class="level3">
<h3 class="anchored" data-anchor-id="batchsvg-implementation"><code>BatchSVG</code> implementation</h3>
<section id="installation" class="level4">
<h4 class="anchored" data-anchor-id="installation">Installation</h4>
<p>(After accepted in <a href="https://bioconductor.org/">Bioconductor</a>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>)) {</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"BatchSVG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Install the development version from <a href="https://christinehou11.github.io/BatchSVG">GitHub</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install</span>(<span class="st">"christinehou11/BatchSVG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="batchsvgfeatureselect-perform-feature-selection" class="level4">
<h4 class="anchored" data-anchor-id="batchsvgfeatureselect-perform-feature-selection"><code>BatchSVG::featureSelect()</code>: Perform Feature Selection</h4>
<p>We first applied <code>featureSelect()</code> function from <code>BatchSVG</code> package to compute the number of standard deviations for the relative change in deviance (<strong>nSD_dev_subject</strong>) and rank difference (<strong>nSD_rank_subject</strong>) for <code>batch = subject</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>list_batch_df <span class="ot">&lt;-</span> <span class="fu">featureSelect</span>(<span class="at">input =</span> spatialLIBD_spe, </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">batch_effect =</span> <span class="st">"subject"</span>, <span class="at">VGs =</span> res_ranks_LIBD<span class="sc">$</span>gene_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Running feature selection without batch...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Batch Effect: subject</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Running feature selection without batch...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating deviance and rank difference...</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(list_batch_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "list"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(list_batch_df<span class="sc">$</span>subject)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default dev_subject rank_subject
1 ENSG00000187608     ISG15    43591.43         1151    43304.39         1157
2 ENSG00000131584     ACAP3    39115.97         1498    38826.05         1509
3 ENSG00000242485    MRPL20    47074.23          920    46912.53          918
4 ENSG00000160075     SSU72    47815.13          862    47742.10          852
5 ENSG00000078369      GNB1    54464.75          477    54385.81          462
6 ENSG00000187730     GABRD    51521.13          646    51099.35          648
       d_diff nSD_dev_subject r_diff nSD_rank_subject
1 0.006628419     -0.04505973      6       0.15937488
2 0.007467317     -0.03697398     11       0.29218728
3 0.003446929     -0.07572463     -2      -0.05312496
4 0.001529688     -0.09420402    -10      -0.26562480
5 0.001451448     -0.09495814    -15      -0.39843720
6 0.008254157     -0.02939000      2       0.05312496</code></pre>
</div>
</div>
</section>
<section id="batchsvgsvg_nsd-visualize-svg-selection-for-batch-effects" class="level4">
<h4 class="anchored" data-anchor-id="batchsvgsvg_nsd-visualize-svg-selection-for-batch-effects"><code>BatchSVG::svg_nSD()</code>: Visualize SVG Selection for Batch Effect(s)</h4>
<p>We utilized <code>svg_nSD()</code> function to generate visualizations for batch effects assessments in spatially variable genes (SVGs).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">svg_nSD</span>(<span class="at">list_batch_df =</span> list_batch_df, </span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">sd_interval_dev =</span> <span class="dv">4</span>, <span class="at">sd_interval_rank =</span> <span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Figure 1. Visualizations of nSD_dev and nSD_rank threshold selection</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>plots<span class="sc">$</span>subject</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="spatialLIBD_spe_files/figure-html/figure 1-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section>
<section id="batchsvgbiasdetect-identify-biased-genes" class="level4">
<h4 class="anchored" data-anchor-id="batchsvgbiasdetect-identify-biased-genes"><code>BatchSVG::biasDetect()</code>: Identify Biased Genes</h4>
<p>Using <code>nSD_svg()</code> function, we determined that <code>nSD_dev = 4</code> and <code>nSD_rank = 6</code>. Then, we can applied <code>biasDetect()</code> function on <code>spe</code> object to identify and filter out biased genes. We used both thresholds (<code>threshold = "both"</code>) and customized the plot color (<code>plot_palette = "RdPu"</code>), the point shape and shape (<code>plot_point_shape = 23</code>), and the text size for biased genes (<code>plot_text_size = 4</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>bias_both <span class="ot">&lt;-</span> <span class="fu">biasDetect</span>(<span class="at">list_batch_df =</span> list_batch_df, <span class="at">threshold =</span> <span class="st">"both"</span>,</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">nSD_dev =</span> <span class="dv">4</span>, <span class="at">nSD_rank =</span> <span class="dv">7</span>, <span class="at">plot_point_shape =</span> <span class="dv">23</span>, <span class="at">plot_palette =</span> <span class="st">"RdPu"</span>,</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot_text_size =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Table 1. Outlier Genes defined by nSD_dev and nSD_rank</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bias_both<span class="sc">$</span>subject<span class="sc">$</span>Table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name dev_default rank_default dev_subject rank_subject
1 ENSG00000255823  MTRNR2L8    286927.9            5    67738.95           77
2 ENSG00000256618  MTRNR2L1    167381.9           14    40299.65         1368
    d_diff nSD_dev_subject r_diff nSD_rank_subject nSD_bin_dev dev_outlier
1 3.235789        31.07931     72         1.912499     [28,32]        TRUE
2 3.153432        30.28551   1354        35.965598     [28,32]        TRUE
  nSD_bin_rank rank_outlier
1        [0,7)        FALSE
2      [35,42]         TRUE</code></pre>
</div>
</div>
<p><em>Figure 2. Outlier Genes defined by nSD_dev and nSD_rank</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>bias_both<span class="sc">$</span>subject<span class="sc">$</span>Plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="spatialLIBD_spe_files/figure-html/figure 2-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section>
</section>
<section id="after-batchsvg-implementation" class="level3">
<h3 class="anchored" data-anchor-id="after-batchsvg-implementation">After <code>BatchSVG</code> implementation</h3>
<section id="subject-related-biased-features" class="level4">
<h4 class="anchored" data-anchor-id="subject-related-biased-features">Subject-related biased features</h4>
<p>Two identified biased features are <strong>MTRNR2L1</strong> and <strong>MTRNR2L8</strong>. Spatial expression plots illustrate that these features display strong subject-biased expression and are also clearly enriched in certain spatial domains. <strong>MTRNR2L1</strong> and <strong>MTRNR2L8</strong> are perfect examples of features that may introduce noise if included in spatial domain clustering.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>ml1 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spatialLIBD_spe, <span class="at">annotate=</span><span class="st">"ENSG00000256618"</span>, <span class="at">assay=</span><span class="st">"logcounts"</span>, </span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">sample_id=</span><span class="st">"sample_id"</span>, <span class="at">point_size=</span>.<span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggtitle</span>(<span class="st">"MTRNR2L1"</span>) <span class="sc">+</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_color_gradient</span>(<span class="at">low=</span><span class="st">'grey90'</span>, <span class="at">high=</span><span class="st">'black'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>ml8 <span class="ot">&lt;-</span> <span class="fu">plotSpots</span>(spatialLIBD_spe, <span class="at">annotate=</span><span class="st">"ENSG00000255823"</span>, <span class="at">assay=</span><span class="st">"logcounts"</span>, </span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">sample_id=</span><span class="st">"sample_id"</span>, <span class="at">point_size=</span>.<span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggtitle</span>(<span class="st">"MTRNR2L8"</span>) <span class="sc">+</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_color_gradient</span>(<span class="at">low=</span><span class="st">'grey90'</span>, <span class="at">high=</span><span class="st">'black'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grid.arrange</span>(ml1, ml8, <span class="at">ncol=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="spatialLIBD_spe_files/figure-html/subject-related plots-1.png" class="img-fluid" width="1344"></p>
</div>
</div>
</section>
<section id="refined-svgs" class="level4">
<h4 class="anchored" data-anchor-id="refined-svgs">Refined SVGs</h4>
<p>Finally, we obtain a refined set of spatially variable genes (SVGs) by removing the identified outliers based on user-defined thresholds for <code>nSD_dev</code> and <code>nSD_rank</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>bias_both_df <span class="ot">&lt;-</span> bias_both<span class="sc">$</span>subject<span class="sc">$</span>Table</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>svgs_filt <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(res_ranks_LIBD<span class="sc">$</span>gene_id, bias_both_df<span class="sc">$</span>gene_id)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>svgs_filt_spe <span class="ot">&lt;-</span> res_ranks_LIBD[res_ranks_LIBD<span class="sc">$</span>gene_id <span class="sc">%in%</span> svgs_filt, ]</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(svgs_filt_spe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1965</code></pre>
</div>
</div>
</section>
</section>
<section id="r-session-information" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="r-session-information"><code>R</code> session information</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Session info</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.2 (2024-10-31)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] gridExtra_2.3               ggspavis_1.12.0            
 [3] BatchSVG_0.99.6             Seurat_5.2.1               
 [5] SeuratObject_5.0.2          sp_2.2-0                   
 [7] PRECAST_1.6.5               gtools_3.9.5               
 [9] here_1.0.1                  scater_1.34.1              
[11] ggplot2_3.5.1               scuttle_1.16.0             
[13] nnSVG_1.10.3                spatialLIBD_1.18.0         
[15] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
[17] SummarizedExperiment_1.36.0 Biobase_2.66.0             
[19] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
[21] IRanges_2.40.1              S4Vectors_0.44.0           
[23] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      
[25] matrixStats_1.5.0          

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.1-0    bitops_1.0-9             httr_1.4.7              
  [4] RColorBrewer_1.1-3       doParallel_1.0.17        backports_1.5.0         
  [7] sctransform_0.4.1        tools_4.4.2              R6_2.6.1                
 [10] DT_0.33                  uwot_0.2.3               lazyeval_0.2.2          
 [13] withr_3.0.2              progressr_0.15.1         cli_3.6.4               
 [16] spatstat.explore_3.4-2   fastDummies_1.7.5        labeling_0.4.3          
 [19] sass_0.4.9               spatstat.data_3.1-6      ggridges_0.5.6          
 [22] pbapply_1.7-2            Rsamtools_2.22.0         parallelly_1.43.0       
 [25] sessioninfo_1.2.3        attempt_0.3.1            maps_3.4.2.1            
 [28] limma_3.62.2             rstudioapi_0.17.1        RSQLite_2.3.9           
 [31] generics_0.1.3           BiocIO_1.16.0            spatstat.random_3.3-3   
 [34] ica_1.0-3                car_3.1-3                dplyr_1.1.4             
 [37] Matrix_1.7-3             ggbeeswarm_0.7.2         abind_1.4-8             
 [40] lifecycle_1.0.4          yaml_2.3.10              edgeR_4.4.2             
 [43] carData_3.0-5            CompQuadForm_1.4.3       SparseArray_1.6.2       
 [46] BiocFileCache_2.14.0     Rtsne_0.17               paletteer_1.6.0         
 [49] grid_4.4.2               blob_1.2.4               promises_1.3.2          
 [52] ExperimentHub_2.14.0     crayon_1.5.3             GiRaF_1.0.1             
 [55] miniUI_0.1.1.1           lattice_0.22-6           beachmat_2.22.0         
 [58] cowplot_1.1.3            KEGGREST_1.46.0          magick_2.8.6            
 [61] pillar_1.10.1            knitr_1.50               rjson_0.2.23            
 [64] future.apply_1.11.3      codetools_0.2-20         glue_1.8.0              
 [67] spatstat.univar_3.1-2    data.table_1.17.0        vctrs_0.6.5             
 [70] png_0.1-8                spam_2.11-1              gtable_0.3.6            
 [73] rematch2_2.1.2           cachem_1.1.0             xfun_0.51               
 [76] S4Arrays_1.6.0           mime_0.13                ggside_0.3.1            
 [79] survival_3.8-3           iterators_1.0.14         fields_16.3.1           
 [82] statmod_1.5.0            fitdistrplus_1.2-2       ROCR_1.0-11             
 [85] nlme_3.1-167             bit64_4.6.0-1            filelock_1.0.3          
 [88] RcppAnnoy_0.0.22         rprojroot_2.0.4          bslib_0.9.0             
 [91] irlba_2.3.5.1            vipor_0.4.7              KernSmooth_2.23-26      
 [94] colorspace_2.1-1         DBI_1.2.3                tidyselect_1.2.1        
 [97] DR.SC_3.4                BRISC_1.0.6              bit_4.6.0               
[100] compiler_4.4.2           curl_6.2.2               BiocNeighbors_2.0.1     
[103] DelayedArray_0.32.0      plotly_4.10.4            rtracklayer_1.66.0      
[106] scales_1.3.0             lmtest_0.9-40            rappdirs_0.3.3          
[109] goftest_1.2-3            stringr_1.5.1            digest_0.6.37           
[112] spatstat.utils_3.1-3     rmarkdown_2.29           benchmarkmeData_1.0.4   
[115] XVector_0.46.0           htmltools_0.5.8.1        pkgconfig_2.0.3         
[118] dbplyr_2.5.0             fastmap_1.2.0            ggthemes_5.1.0          
[121] rlang_1.1.5              htmlwidgets_1.6.4        UCSC.utils_1.2.0        
[124] scry_1.18.0              shiny_1.10.0             farver_2.1.2            
[127] jquerylib_0.1.4          zoo_1.8-13               jsonlite_1.9.1          
[130] BiocParallel_1.40.0      mclust_6.1.1             config_0.3.2            
[133] BiocSingular_1.22.0      RCurl_1.98-1.17          magrittr_2.0.3          
[136] Formula_1.2-5            GenomeInfoDbData_1.2.13  dotCall64_1.2           
[139] patchwork_1.3.0          munsell_0.5.1            Rcpp_1.0.14             
[142] reticulate_1.41.0.1      viridis_0.6.5            stringi_1.8.4           
[145] zlibbioc_1.52.0          MASS_7.3-65              plyr_1.8.9              
[148] AnnotationHub_3.14.0     listenv_0.9.1            ggrepel_0.9.6           
[151] deldir_2.0-4             Biostrings_2.74.1        splines_4.4.2           
[154] tensor_1.5               locfit_1.5-9.12          rdist_0.0.5             
[157] ggpubr_0.6.0             igraph_2.1.4             spatstat.geom_3.3-6     
[160] ggsignif_0.6.4           RcppHNSW_0.6.0           reshape2_1.4.4          
[163] ScaledMatrix_1.14.0      BiocVersion_3.20.0       XML_3.99-0.18           
[166] evaluate_1.0.3           golem_0.5.1              BiocManager_1.30.25     
[169] foreach_1.5.2            httpuv_1.6.15            RANN_2.6.2              
[172] tidyr_1.3.1              purrr_1.0.4              polyclip_1.10-7         
[175] scattermore_1.2          future_1.34.0            benchmarkme_1.0.8       
[178] rsvd_1.0.5               broom_1.0.7              xtable_1.8-4            
[181] restfulr_0.0.15          RSpectra_0.16-2          rstatix_0.7.2           
[184] later_1.4.1              viridisLite_0.4.2        tibble_3.2.1            
[187] memoise_2.0.1            beeswarm_0.4.0           AnnotationDbi_1.68.0    
[190] GenomicAlignments_1.42.0 cluster_2.1.8.1          shinyWidgets_0.9.0      
[193] globals_0.16.3          </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>